<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Methods | Supplementary Information: Cellxgene VIP unleashes full power of interactive visualization and integrative analysis of scRNA-seq, spatial transcriptomics, and multiome data</title>
  <meta name="description" content="Chapter 3 Methods | Supplementary Information: Cellxgene VIP unleashes full power of interactive visualization and integrative analysis of scRNA-seq, spatial transcriptomics, and multiome data" />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Methods | Supplementary Information: Cellxgene VIP unleashes full power of interactive visualization and integrative analysis of scRNA-seq, spatial transcriptomics, and multiome data" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Methods | Supplementary Information: Cellxgene VIP unleashes full power of interactive visualization and integrative analysis of scRNA-seq, spatial transcriptomics, and multiome data" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="how-to-use-cellxgene-vip.html"/>
<link rel="next" href="helpful-tips.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">CellXgene VIP</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Getting started with cellxgene VIP</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#why-use-cellxgene-vip"><i class="fa fa-check"></i><b>1.1</b> Why use cellxgene VIP?</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#getting-set-up"><i class="fa fa-check"></i><b>1.2</b> Getting Set up</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="index.html"><a href="index.html#execute-anaconda"><i class="fa fa-check"></i><b>1.2.1</b> Execute anaconda</a></li>
<li class="chapter" data-level="1.2.2" data-path="index.html"><a href="index.html#cellxgene-installation"><i class="fa fa-check"></i><b>1.2.2</b> Cellxgene installation</a></li>
<li class="chapter" data-level="1.2.3" data-path="index.html"><a href="index.html#r-dependencies"><i class="fa fa-check"></i><b>1.2.3</b> R dependencies</a></li>
<li class="chapter" data-level="1.2.4" data-path="index.html"><a href="index.html#run-cellxgene-by-h5ad-file"><i class="fa fa-check"></i><b>1.2.4</b> Run cellxgene by h5ad file</a></li>
<li class="chapter" data-level="1.2.5" data-path="index.html"><a href="index.html#cellxgene-on-web-browser"><i class="fa fa-check"></i><b>1.2.5</b> Cellxgene on web browser</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#authors"><i class="fa fa-check"></i><b>1.3</b> Authors</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html"><i class="fa fa-check"></i><b>2</b> How to use Cellxgene VIP</a>
<ul>
<li class="chapter" data-level="2.1" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#graphical-user-interface-of-cellxgene-and-vip"><i class="fa fa-check"></i><b>2.1</b> Graphical user interface of cellxgene and VIP</a></li>
<li class="chapter" data-level="2.2" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#cell-selection-by-categorial-annotations"><i class="fa fa-check"></i><b>2.2</b> Cell selection by categorial annotations</a></li>
<li class="chapter" data-level="2.3" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#cell-selection-by-brushing-on-distribution-of-continues-variables"><i class="fa fa-check"></i><b>2.3</b> Cell selection by brushing on distribution of continues variables</a></li>
<li class="chapter" data-level="2.4" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#free-hand-lasso-selection-on-dots-representing-cells"><i class="fa fa-check"></i><b>2.4</b> Free hand Lasso selection on dots representing cells</a></li>
<li class="chapter" data-level="2.5" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-global-setting"><i class="fa fa-check"></i><b>2.5</b> VIP – Global Setting</a></li>
<li class="chapter" data-level="2.6" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-add-genes-gene-sets"><i class="fa fa-check"></i><b>2.6</b> VIP – Add Genes / Gene Sets</a></li>
<li class="chapter" data-level="2.7" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-violin-plot"><i class="fa fa-check"></i><b>2.7</b> VIP – Violin Plot</a></li>
<li class="chapter" data-level="2.8" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-heatmap"><i class="fa fa-check"></i><b>2.8</b> VIP – Heatmap</a></li>
<li class="chapter" data-level="2.9" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-embedding"><i class="fa fa-check"></i><b>2.9</b> VIP – Embedding</a></li>
<li class="chapter" data-level="2.10" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-dot-plot"><i class="fa fa-check"></i><b>2.10</b> VIP – Dot Plot</a></li>
<li class="chapter" data-level="2.11" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-track-plot"><i class="fa fa-check"></i><b>2.11</b> VIP – Track Plot</a></li>
<li class="chapter" data-level="2.12" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-density-plot"><i class="fa fa-check"></i><b>2.12</b> VIP – Density Plot</a></li>
<li class="chapter" data-level="2.13" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-density-scatter-plot"><i class="fa fa-check"></i><b>2.13</b> VIP – Density Scatter Plot</a></li>
<li class="chapter" data-level="2.14" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-dual-genes"><i class="fa fa-check"></i><b>2.14</b> VIP – Dual Genes</a></li>
<li class="chapter" data-level="2.15" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-sankey-diagram"><i class="fa fa-check"></i><b>2.15</b> VIP – Sankey Diagram</a></li>
<li class="chapter" data-level="2.16" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-stacked-barplot"><i class="fa fa-check"></i><b>2.16</b> VIP – Stacked Barplot</a></li>
<li class="chapter" data-level="2.17" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-gene-specificity"><i class="fa fa-check"></i><b>2.17</b> VIP – Gene Specificity</a></li>
<li class="chapter" data-level="2.18" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-gene-detected"><i class="fa fa-check"></i><b>2.18</b> VIP – Gene Detected</a></li>
<li class="chapter" data-level="2.19" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-deg-differential-expressed-genes"><i class="fa fa-check"></i><b>2.19</b> VIP – DEG (Differential Expressed Genes)</a></li>
<li class="chapter" data-level="2.20" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip---pre-computed-deg"><i class="fa fa-check"></i><b>2.20</b> VIP - Pre-computed DEG</a></li>
<li class="chapter" data-level="2.21" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-marker-genes"><i class="fa fa-check"></i><b>2.21</b> VIP – Marker Genes</a></li>
<li class="chapter" data-level="2.22" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip---cosmx"><i class="fa fa-check"></i><b>2.22</b> VIP - CosMx</a></li>
<li class="chapter" data-level="2.23" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip---spatial-transcriptomics"><i class="fa fa-check"></i><b>2.23</b> VIP - Spatial Transcriptomics</a></li>
<li class="chapter" data-level="2.24" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-command-line-interface"><i class="fa fa-check"></i><b>2.24</b> VIP – Command Line Interface</a></li>
<li class="chapter" data-level="2.25" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-comb.-abbr."><i class="fa fa-check"></i><b>2.25</b> VIP – Comb. &amp; Abbr.</a></li>
<li class="chapter" data-level="2.26" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-other-functions"><i class="fa fa-check"></i><b>2.26</b> VIP – Other Functions</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="methods.html"><a href="methods.html"><i class="fa fa-check"></i><b>3</b> Methods</a>
<ul>
<li class="chapter" data-level="3.1" data-path="methods.html"><a href="methods.html#client-side-integration-by-a-jspanel-window-vip"><i class="fa fa-check"></i><b>3.1</b> Client-side Integration by a jsPanel Window (VIP)</a></li>
<li class="chapter" data-level="3.2" data-path="methods.html"><a href="methods.html#server-side-integration"><i class="fa fa-check"></i><b>3.2</b> Server-side Integration</a></li>
<li class="chapter" data-level="3.3" data-path="methods.html"><a href="methods.html#communication-between-vip-and-cellxgene-web-gui"><i class="fa fa-check"></i><b>3.3</b> Communication between VIP and cellxgene web GUI</a></li>
<li class="chapter" data-level="3.4" data-path="methods.html"><a href="methods.html#diffxpy-integration"><i class="fa fa-check"></i><b>3.4</b> Diffxpy Integration</a></li>
<li class="chapter" data-level="3.5" data-path="methods.html"><a href="methods.html#create-a-h5ad-file-from-seurat-object"><i class="fa fa-check"></i><b>3.5</b> Create a h5ad file from Seurat object</a></li>
<li class="chapter" data-level="3.6" data-path="methods.html"><a href="methods.html#prepare-spatial-data-for-visualization"><i class="fa fa-check"></i><b>3.6</b> Prepare Spatial Data for Visualization</a>
<ul>
<li class="chapter" data-level="3.6.1" data-path="methods.html"><a href="methods.html#x-visium-data"><i class="fa fa-check"></i><b>3.6.1</b> 10X visium data</a></li>
<li class="chapter" data-level="3.6.2" data-path="methods.html"><a href="methods.html#nanostring-cosmx-data"><i class="fa fa-check"></i><b>3.6.2</b> NanoString CosMx data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="helpful-tips.html"><a href="helpful-tips.html"><i class="fa fa-check"></i><b>4</b> Helpful Tips</a>
<ul>
<li class="chapter" data-level="4.1" data-path="helpful-tips.html"><a href="helpful-tips.html#handle-nulls-in-categorical-annotation"><i class="fa fa-check"></i><b>4.1</b> Handle nulls in categorical annotation</a></li>
<li class="chapter" data-level="4.2" data-path="helpful-tips.html"><a href="helpful-tips.html#display-full-traceback-stack-for-debugging-in-vip"><i class="fa fa-check"></i><b>4.2</b> Display full traceback stack for debugging in VIP</a></li>
<li class="chapter" data-level="4.3" data-path="helpful-tips.html"><a href="helpful-tips.html#pitfall-of-using-special-characters"><i class="fa fa-check"></i><b>4.3</b> Pitfall of using special characters</a></li>
<li class="chapter" data-level="4.4" data-path="helpful-tips.html"><a href="helpful-tips.html#potential-use-for-bulk-or-pseudo-bulk-sample-dataset"><i class="fa fa-check"></i><b>4.4</b> Potential use for bulk or pseudo bulk sample dataset</a></li>
<li class="chapter" data-level="4.5" data-path="helpful-tips.html"><a href="helpful-tips.html#common-mistakes-in-naming"><i class="fa fa-check"></i><b>4.5</b> Common mistakes in naming</a></li>
<li class="chapter" data-level="4.6" data-path="helpful-tips.html"><a href="helpful-tips.html#loading-h5ad-file-generated-by-seuratdisk-from-r"><i class="fa fa-check"></i><b>4.6</b> Loading h5ad file generated by SeuratDisk from R</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="web-resource.html"><a href="web-resource.html"><i class="fa fa-check"></i><b>5</b> Web Resource</a>
<ul>
<li class="chapter" data-level="5.1" data-path="web-resource.html"><a href="web-resource.html#cellxgene-vip"><i class="fa fa-check"></i><b>5.1</b> Cellxgene VIP</a></li>
<li class="chapter" data-level="5.2" data-path="web-resource.html"><a href="web-resource.html#cellxgene"><i class="fa fa-check"></i><b>5.2</b> Cellxgene</a></li>
<li class="chapter" data-level="5.3" data-path="web-resource.html"><a href="web-resource.html#python-packages"><i class="fa fa-check"></i><b>5.3</b> Python packages</a></li>
<li class="chapter" data-level="5.4" data-path="web-resource.html"><a href="web-resource.html#others"><i class="fa fa-check"></i><b>5.4</b> Others</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Supplementary Information: Cellxgene VIP unleashes full power of interactive visualization and integrative analysis of scRNA-seq, spatial transcriptomics, and multiome data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="methods" class="section level1 hasAnchor" number="3">
<h1><span class="header-section-number">Chapter 3</span> Methods<a href="methods.html#methods" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="client-side-integration-by-a-jspanel-window-vip" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> Client-side Integration by a jsPanel Window (VIP)<a href="methods.html#client-side-integration-by-a-jspanel-window-vip" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Related lines in config.sh file.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="methods.html#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-i</span> <span class="st">&quot;s|&lt;div id=</span><span class="dt">\&quot;</span><span class="st">root</span><span class="dt">\&quot;</span><span class="st">&gt;&lt;/div&gt;|</span><span class="va">$(</span><span class="fu">sed</span> <span class="at">-e</span> <span class="st">&#39;s/[&amp;\\/]/\\&amp;/g; s/|/\\|/g; s/$/\\/;&#39;</span> <span class="at">-e</span> <span class="st">&#39;$s/\\$//&#39;</span> index_template.insert<span class="va">)</span><span class="st">\n&amp;|&quot;</span> <span class="st">&quot;cellxgene/client/index_template.html&quot;</span></span></code></pre></div>
<p>The content of the index_template.insert file.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode js"><code class="sourceCode javascript"><span id="cb9-1"><a href="methods.html#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>script src<span class="op">=</span><span class="st">&quot;https://interactivereport.github.io/cellxgene_VIP/static/jquery.min.js&quot;</span><span class="op">&gt;&lt;/</span>script<span class="op">&gt;</span></span>
<span id="cb9-2"><a href="methods.html#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>script src<span class="op">=</span><span class="st">&quot;https://d3js.org/d3.v4.min.js&quot;</span><span class="op">&gt;&lt;/</span>script<span class="op">&gt;</span></span>
<span id="cb9-3"><a href="methods.html#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>script src<span class="op">=</span><span class="st">&quot;https://interactivereport.github.io/cellxgene_VIP/static/stackedbar/d3.v3.min.js&quot;</span><span class="op">&gt;&lt;/</span>script<span class="op">&gt;</span></span>
<span id="cb9-4"><a href="methods.html#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>link href<span class="op">=</span><span class="st">&quot;https://interactivereport.github.io/cellxgene_VIP/static/jspanel/dist/jspanel.css&quot;</span> rel<span class="op">=</span><span class="st">&quot;stylesheet&quot;</span><span class="op">&gt;</span></span>
<span id="cb9-5"><a href="methods.html#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>script src<span class="op">=</span><span class="st">&quot;https://interactivereport.github.io/cellxgene_VIP/static/jspanel/dist/jspanel.js&quot;</span><span class="op">&gt;&lt;/</span>script<span class="op">&gt;</span></span>
<span id="cb9-6"><a href="methods.html#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>script src<span class="op">=</span><span class="st">&quot;https://interactivereport.github.io/cellxgene_VIP/static/jspanel/dist/extensions/modal/jspanel.modal.js&quot;</span><span class="op">&gt;&lt;/</span>script<span class="op">&gt;</span></span>
<span id="cb9-7"><a href="methods.html#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>script src<span class="op">=</span><span class="st">&quot;https://interactivereport.github.io/cellxgene_VIP/static/jspanel/dist/extensions/tooltip/jspanel.tooltip.js&quot;</span><span class="op">&gt;&lt;/</span>script<span class="op">&gt;</span></span>
<span id="cb9-8"><a href="methods.html#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>script src<span class="op">=</span><span class="st">&quot;https://interactivereport.github.io/cellxgene_VIP/static/jspanel/dist/extensions/hint/jspanel.hint.js&quot;</span><span class="op">&gt;&lt;/</span>script<span class="op">&gt;</span></span>
<span id="cb9-9"><a href="methods.html#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>script src<span class="op">=</span><span class="st">&quot;https://interactivereport.github.io/cellxgene_VIP/static/jspanel/dist/extensions/layout/jspanel.layout.js&quot;</span><span class="op">&gt;&lt;/</span>script<span class="op">&gt;</span></span>
<span id="cb9-10"><a href="methods.html#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>script src<span class="op">=</span><span class="st">&quot;https://interactivereport.github.io/cellxgene_VIP/static/jspanel/dist/extensions/contextmenu/jspanel.contextmenu.js&quot;</span><span class="op">&gt;&lt;/</span>script<span class="op">&gt;</span></span>
<span id="cb9-11"><a href="methods.html#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>script src<span class="op">=</span><span class="st">&quot;https://interactivereport.github.io/cellxgene_VIP/static/jspanel/dist/extensions/dock/jspanel.dock.js&quot;</span><span class="op">&gt;&lt;/</span>script<span class="op">&gt;</span></span>
<span id="cb9-12"><a href="methods.html#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>script<span class="op">&gt;</span></span>
<span id="cb9-13"><a href="methods.html#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="co">// execute JavaScript code in panel content</span></span>
<span id="cb9-14"><a href="methods.html#cb9-14" aria-hidden="true" tabindex="-1"></a><span class="kw">var</span> setInnerHTML <span class="op">=</span> <span class="kw">function</span>(elm<span class="op">,</span> html) {</span>
<span id="cb9-15"><a href="methods.html#cb9-15" aria-hidden="true" tabindex="-1"></a>    elm<span class="op">.</span><span class="at">innerHTML</span> <span class="op">=</span> html<span class="op">;</span></span>
<span id="cb9-16"><a href="methods.html#cb9-16" aria-hidden="true" tabindex="-1"></a>    <span class="bu">Array</span><span class="op">.</span><span class="fu">from</span>(elm<span class="op">.</span><span class="fu">querySelectorAll</span>(<span class="st">&#39;script&#39;</span>))<span class="op">.</span><span class="fu">forEach</span>( oldScript <span class="kw">=&gt;</span> {</span>
<span id="cb9-17"><a href="methods.html#cb9-17" aria-hidden="true" tabindex="-1"></a>        <span class="kw">const</span> newScript <span class="op">=</span> <span class="bu">document</span><span class="op">.</span><span class="fu">createElement</span>(<span class="st">&#39;script&#39;</span>)<span class="op">;</span></span>
<span id="cb9-18"><a href="methods.html#cb9-18" aria-hidden="true" tabindex="-1"></a>        <span class="bu">Array</span><span class="op">.</span><span class="fu">from</span>(oldScript<span class="op">.</span><span class="at">attributes</span>)</span>
<span id="cb9-19"><a href="methods.html#cb9-19" aria-hidden="true" tabindex="-1"></a>        <span class="op">.</span><span class="fu">forEach</span>( attr <span class="kw">=&gt;</span> newScript<span class="op">.</span><span class="fu">setAttribute</span>(attr<span class="op">.</span><span class="at">name</span><span class="op">,</span> attr<span class="op">.</span><span class="at">value</span>) )<span class="op">;</span></span>
<span id="cb9-20"><a href="methods.html#cb9-20" aria-hidden="true" tabindex="-1"></a>        newScript<span class="op">.</span><span class="fu">appendChild</span>(<span class="bu">document</span><span class="op">.</span><span class="fu">createTextNode</span>(oldScript<span class="op">.</span><span class="at">innerHTML</span>))<span class="op">;</span></span>
<span id="cb9-21"><a href="methods.html#cb9-21" aria-hidden="true" tabindex="-1"></a>        oldScript<span class="op">.</span><span class="at">parentNode</span><span class="op">.</span><span class="fu">replaceChild</span>(newScript<span class="op">,</span> oldScript)<span class="op">;</span></span>
<span id="cb9-22"><a href="methods.html#cb9-22" aria-hidden="true" tabindex="-1"></a>    })<span class="op">;</span></span>
<span id="cb9-23"><a href="methods.html#cb9-23" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb9-24"><a href="methods.html#cb9-24" aria-hidden="true" tabindex="-1"></a><span class="kw">var</span> plotPanel <span class="op">=</span> jsPanel<span class="op">.</span><span class="fu">create</span>({</span>
<span id="cb9-25"><a href="methods.html#cb9-25" aria-hidden="true" tabindex="-1"></a>    <span class="dt">panelSize</span><span class="op">:</span> <span class="st">&#39;190 0&#39;</span><span class="op">,</span></span>
<span id="cb9-26"><a href="methods.html#cb9-26" aria-hidden="true" tabindex="-1"></a>    <span class="dt">position</span><span class="op">:</span> <span class="st">&#39;left-top 160 6&#39;</span><span class="op">,</span></span>
<span id="cb9-27"><a href="methods.html#cb9-27" aria-hidden="true" tabindex="-1"></a>    <span class="dt">dragit</span><span class="op">:</span> { <span class="dt">containment</span><span class="op">:</span> [<span class="op">-</span><span class="dv">10</span><span class="op">,</span> <span class="op">-</span><span class="dv">2000</span><span class="op">,</span> <span class="op">-</span><span class="dv">4000</span><span class="op">,</span> <span class="op">-</span><span class="dv">2000</span>] }<span class="op">,</span> <span class="co">// set dragging range of VIP window</span></span>
<span id="cb9-28"><a href="methods.html#cb9-28" aria-hidden="true" tabindex="-1"></a>    <span class="dt">boxShadow</span><span class="op">:</span> <span class="dv">1</span><span class="op">,</span></span>
<span id="cb9-29"><a href="methods.html#cb9-29" aria-hidden="true" tabindex="-1"></a>    <span class="dt">border</span><span class="op">:</span> <span class="st">&quot;solid #D4DBDE thin&quot;</span><span class="op">,</span></span>
<span id="cb9-30"><a href="methods.html#cb9-30" aria-hidden="true" tabindex="-1"></a>    <span class="dt">contentOverflow</span><span class="op">:</span> <span class="st">&#39;scroll scroll&#39;</span><span class="op">,</span> <span class="co">// adding scrolling bars</span></span>
<span id="cb9-31"><a href="methods.html#cb9-31" aria-hidden="true" tabindex="-1"></a>    <span class="dt">headerControls</span><span class="op">:</span>{</span>
<span id="cb9-32"><a href="methods.html#cb9-32" aria-hidden="true" tabindex="-1"></a>      <span class="dt">close</span><span class="op">:</span> <span class="st">&#39;remove&#39;</span><span class="op">,</span></span>
<span id="cb9-33"><a href="methods.html#cb9-33" aria-hidden="true" tabindex="-1"></a>      <span class="dt">minimize</span><span class="op">:</span> <span class="st">&#39;remove&#39;</span><span class="op">,</span></span>
<span id="cb9-34"><a href="methods.html#cb9-34" aria-hidden="true" tabindex="-1"></a>      <span class="dt">maximize</span><span class="op">:</span> <span class="st">&#39;remove&#39;</span></span>
<span id="cb9-35"><a href="methods.html#cb9-35" aria-hidden="true" tabindex="-1"></a>    }<span class="op">,</span></span>
<span id="cb9-36"><a href="methods.html#cb9-36" aria-hidden="true" tabindex="-1"></a>    <span class="dt">headerTitle</span><span class="op">:</span> <span class="kw">function</span> () {<span class="cf">return</span> <span class="st">&#39;&lt;strong&gt;Visualization in Plugin&lt;/strong&gt;&#39;</span>}<span class="op">,</span></span>
<span id="cb9-37"><a href="methods.html#cb9-37" aria-hidden="true" tabindex="-1"></a>    <span class="dt">contentAjax</span><span class="op">:</span> {</span>
<span id="cb9-38"><a href="methods.html#cb9-38" aria-hidden="true" tabindex="-1"></a>        <span class="dt">url</span><span class="op">:</span> <span class="bu">window</span><span class="op">.</span><span class="at">location</span><span class="op">.</span><span class="at">href</span><span class="op">.</span><span class="fu">replace</span>(<span class="ss">/</span><span class="sc">\\\/+$</span><span class="ss">/</span><span class="op">,</span><span class="st">&#39;&#39;</span>)<span class="op">+</span><span class="st">&#39;/static/interface.html&#39;</span><span class="op">,</span></span>
<span id="cb9-39"><a href="methods.html#cb9-39" aria-hidden="true" tabindex="-1"></a>        <span class="dt">done</span><span class="op">:</span> <span class="kw">function</span> (panel) {</span>
<span id="cb9-40"><a href="methods.html#cb9-40" aria-hidden="true" tabindex="-1"></a>               <span class="fu">setInnerHTML</span>(panel<span class="op">.</span><span class="at">content</span><span class="op">,</span> <span class="kw">this</span><span class="op">.</span><span class="at">responseText</span>)<span class="op">;</span></span>
<span id="cb9-41"><a href="methods.html#cb9-41" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb9-42"><a href="methods.html#cb9-42" aria-hidden="true" tabindex="-1"></a>    }<span class="op">,</span></span>
<span id="cb9-43"><a href="methods.html#cb9-43" aria-hidden="true" tabindex="-1"></a>    <span class="dt">onwindowresize</span><span class="op">:</span> <span class="kw">function</span>(<span class="bu">event</span><span class="op">,</span> panel) {</span>
<span id="cb9-44"><a href="methods.html#cb9-44" aria-hidden="true" tabindex="-1"></a>        <span class="kw">var</span> jptop <span class="op">=</span> <span class="pp">parseInt</span>(<span class="kw">this</span><span class="op">.</span><span class="at">currentData</span><span class="op">.</span><span class="at">top</span>)<span class="op">;</span></span>
<span id="cb9-45"><a href="methods.html#cb9-45" aria-hidden="true" tabindex="-1"></a>        <span class="kw">var</span> jpleft <span class="op">=</span> <span class="pp">parseInt</span>(<span class="kw">this</span><span class="op">.</span><span class="at">currentData</span><span class="op">.</span><span class="at">left</span>)<span class="op">;</span></span>
<span id="cb9-46"><a href="methods.html#cb9-46" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (jptop<span class="op">&lt;-</span><span class="dv">10</span> <span class="op">||</span> <span class="bu">window</span><span class="op">.</span><span class="at">innerHeight</span><span class="op">-</span>jptop<span class="op">&lt;</span><span class="dv">10</span> <span class="op">||</span> <span class="bu">window</span><span class="op">.</span><span class="at">innerWidth</span><span class="op">-</span>jpleft<span class="op">&lt;</span><span class="dv">10</span> <span class="op">||</span> jpleft<span class="op">+</span><span class="pp">parseInt</span>(<span class="kw">this</span><span class="op">.</span><span class="at">currentData</span><span class="op">.</span><span class="at">width</span>)<span class="op">&lt;</span><span class="dv">10</span>) {</span>
<span id="cb9-47"><a href="methods.html#cb9-47" aria-hidden="true" tabindex="-1"></a>            <span class="kw">this</span><span class="op">.</span><span class="fu">reposition</span>(<span class="st">&quot;left-top 160 6&quot;</span>)<span class="op">;</span></span>
<span id="cb9-48"><a href="methods.html#cb9-48" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb9-49"><a href="methods.html#cb9-49" aria-hidden="true" tabindex="-1"></a>    }<span class="op">,</span></span>
<span id="cb9-50"><a href="methods.html#cb9-50" aria-hidden="true" tabindex="-1"></a>    <span class="dt">onunsmallified</span><span class="op">:</span> <span class="kw">function</span> (panel<span class="op">,</span> status) {</span>
<span id="cb9-51"><a href="methods.html#cb9-51" aria-hidden="true" tabindex="-1"></a>        <span class="kw">this</span><span class="op">.</span><span class="fu">reposition</span>(<span class="st">&#39;center-top -370 180&#39;</span>)<span class="op">;</span></span>
<span id="cb9-52"><a href="methods.html#cb9-52" aria-hidden="true" tabindex="-1"></a>        <span class="kw">this</span><span class="op">.</span><span class="fu">resize</span>({ <span class="dt">width</span><span class="op">:</span> <span class="dv">740</span><span class="op">,</span> <span class="dt">height</span><span class="op">:</span> <span class="kw">function</span>() { <span class="cf">return</span> <span class="bu">Math</span><span class="op">.</span><span class="fu">min</span>(<span class="dv">480</span><span class="op">,</span> <span class="bu">window</span><span class="op">.</span><span class="at">innerHeight</span><span class="op">*</span><span class="fl">0.6</span>)<span class="op">;</span>} })<span class="op">;</span></span>
<span id="cb9-53"><a href="methods.html#cb9-53" aria-hidden="true" tabindex="-1"></a>    }<span class="op">,</span></span>
<span id="cb9-54"><a href="methods.html#cb9-54" aria-hidden="true" tabindex="-1"></a>    <span class="dt">onsmallified</span><span class="op">:</span> <span class="kw">function</span> (panel<span class="op">,</span> status) {</span>
<span id="cb9-55"><a href="methods.html#cb9-55" aria-hidden="true" tabindex="-1"></a>        <span class="kw">this</span><span class="op">.</span><span class="fu">reposition</span>(<span class="st">&#39;left-top 160 6&#39;</span>)<span class="op">;</span></span>
<span id="cb9-56"><a href="methods.html#cb9-56" aria-hidden="true" tabindex="-1"></a>        <span class="kw">this</span><span class="op">.</span><span class="at">style</span><span class="op">.</span><span class="at">width</span> <span class="op">=</span> <span class="st">&#39;190px&#39;</span><span class="op">;</span></span>
<span id="cb9-57"><a href="methods.html#cb9-57" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb9-58"><a href="methods.html#cb9-58" aria-hidden="true" tabindex="-1"></a>})<span class="op">.</span><span class="fu">smallify</span>()<span class="op">;</span></span>
<span id="cb9-59"><a href="methods.html#cb9-59" aria-hidden="true" tabindex="-1"></a>plotPanel<span class="op">.</span><span class="at">headerbar</span><span class="op">.</span><span class="at">style</span><span class="op">.</span><span class="at">background</span> <span class="op">=</span> <span class="st">&quot;#D4DBDE&quot;</span><span class="op">;</span></span>
<span id="cb9-60"><a href="methods.html#cb9-60" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;/</span>script<span class="op">&gt;</span></span></code></pre></div>
<p>All functional VIP HTML and JavaScript code will be in a new file called “interface.html”, which is out of cellxgene code base.</p>
</div>
<div id="server-side-integration" class="section level2 hasAnchor" number="3.2">
<h2><span class="header-section-number">3.2</span> Server-side Integration<a href="methods.html#server-side-integration" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Related in config.sh file.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="methods.html#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&#39;</span></span>
<span id="cb10-2"><a href="methods.html#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="st">from server.app.VIPInterface import route</span></span>
<span id="cb10-3"><a href="methods.html#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="st">@webbp.route(&quot;/VIP&quot;, methods=[&quot;POST&quot;])</span></span>
<span id="cb10-4"><a href="methods.html#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="st">def VIP():</span></span>
<span id="cb10-5"><a href="methods.html#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="st">    return route(request.data,current_app.app_config)&#39;</span> <span class="op">&gt;&gt;</span> cellxgene/server/app/app.py</span>
<span id="cb10-6"><a href="methods.html#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="methods.html#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> cellxgene</span>
<span id="cb10-8"><a href="methods.html#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="fu">make</span> pydist</span>
<span id="cb10-9"><a href="methods.html#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="fu">make</span> install-dist</span>
<span id="cb10-10"><a href="methods.html#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb10-11"><a href="methods.html#cb10-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-12"><a href="methods.html#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="co">## finished setting up ------</span></span>
<span id="cb10-13"><a href="methods.html#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="ex">./update.VIPInterface.sh</span> all</span></code></pre></div>
<p>The content of update.VIPInterface.sh file.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="methods.html#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/usr/bin/env bash</span></span>
<span id="cb11-2"><a href="methods.html#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="bu">[</span> <span class="ot">-n</span> <span class="st">&quot;</span><span class="va">$1</span><span class="st">&quot;</span> <span class="bu">]</span><span class="kw">;</span> <span class="cf">then</span></span>
<span id="cb11-3"><a href="methods.html#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;usually update once&quot;</span></span>
<span id="cb11-4"><a href="methods.html#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="cf">fi</span></span>
<span id="cb11-5"><a href="methods.html#cb11-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-6"><a href="methods.html#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="co">## finished setting up ------</span></span>
<span id="cb11-7"><a href="methods.html#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="va">strPath</span><span class="op">=</span><span class="st">&quot;</span><span class="va">$(</span><span class="ex">python</span> <span class="at">-c</span> <span class="st">&#39;import site; print(site.getsitepackages()[0])&#39;</span><span class="va">)</span><span class="st">&quot;</span></span>
<span id="cb11-8"><a href="methods.html#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="va">strweb</span><span class="op">=</span><span class="st">&quot;</span><span class="va">${strPath}</span><span class="st">/server/common/web/static/.&quot;</span></span>
<span id="cb11-9"><a href="methods.html#cb11-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-10"><a href="methods.html#cb11-10" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> VIPInterface.py <span class="va">$strPath</span>/server/app/.</span>
<span id="cb11-11"><a href="methods.html#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> interface.html <span class="va">$strweb</span></span>
<span id="cb11-12"><a href="methods.html#cb11-12" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> vip.env <span class="va">$strPath</span>/server/app/. <span class="dv">2</span><span class="op">&gt;</span>/dev/null <span class="kw">|</span> <span class="fu">true</span></span>
<span id="cb11-13"><a href="methods.html#cb11-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-14"><a href="methods.html#cb11-14" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> fgsea.R <span class="va">$strPath</span>/server/app/.</span>
<span id="cb11-15"><a href="methods.html#cb11-15" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$strPath</span>/server/app/gsea</span>
<span id="cb11-16"><a href="methods.html#cb11-16" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> gsea/<span class="pp">*</span>gmt <span class="va">$strPath</span>/server/app/gsea</span>
<span id="cb11-17"><a href="methods.html#cb11-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-18"><a href="methods.html#cb11-18" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="bu">[</span> <span class="ot">-n</span> <span class="st">&quot;</span><span class="va">$1</span><span class="st">&quot;</span> <span class="bu">]</span><span class="kw">;</span> <span class="cf">then</span></span>
<span id="cb11-19"><a href="methods.html#cb11-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cp</span> Density2D.R <span class="va">$strPath</span>/server/app/.</span>
<span id="cb11-20"><a href="methods.html#cb11-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cp</span> bubbleMap.R <span class="va">$strPath</span>/server/app/.</span>
<span id="cb11-21"><a href="methods.html#cb11-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cp</span> violin.R <span class="va">$strPath</span>/server/app/.</span>
<span id="cb11-22"><a href="methods.html#cb11-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cp</span> volcano.R <span class="va">$strPath</span>/server/app/.</span>
<span id="cb11-23"><a href="methods.html#cb11-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cp</span> browserPlot.R <span class="va">$strPath</span>/server/app/.</span>
<span id="cb11-24"><a href="methods.html#cb11-24" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> <span class="bu">[</span> <span class="st">&quot;</span><span class="va">$(</span><span class="fu">uname</span> <span class="at">-s</span><span class="va">)</span><span class="st">&quot;</span> <span class="ot">=</span> <span class="st">&quot;Darwin&quot;</span> <span class="bu">]</span><span class="kw">;</span> <span class="cf">then</span></span>
<span id="cb11-25"><a href="methods.html#cb11-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">sed</span> <span class="at">-i</span> .bak <span class="st">&quot;s|route(request.data,current_app.app_config, </span><span class="dt">\&quot;</span><span class="st">/tmp</span><span class="dt">\&quot;</span><span class="st">)|route(request.data,current_app.app_config)|&quot;</span> <span class="st">&quot;</span><span class="va">$strPath</span><span class="st">/server/app/app.py&quot;</span></span>
<span id="cb11-26"><a href="methods.html#cb11-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">sed</span> <span class="at">-i</span> .bak <span class="st">&quot;s|MAX_LAYOUTS *= *[0-9]\+|MAX_LAYOUTS = 300|&quot;</span> <span class="st">&quot;</span><span class="va">$strPath</span><span class="st">/server/common/constants.py&quot;</span></span>
<span id="cb11-27"><a href="methods.html#cb11-27" aria-hidden="true" tabindex="-1"></a>  <span class="cf">else</span></span>
<span id="cb11-28"><a href="methods.html#cb11-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">sed</span> <span class="at">-i</span> <span class="st">&quot;s|route(request.data,current_app.app_config, </span><span class="dt">\&quot;</span><span class="st">/tmp</span><span class="dt">\&quot;</span><span class="st">)|route(request.data,current_app.app_config)|&quot;</span> <span class="st">&quot;</span><span class="va">$strPath</span><span class="st">/server/app/app.py&quot;</span></span>
<span id="cb11-29"><a href="methods.html#cb11-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">sed</span> <span class="at">-i</span> <span class="st">&quot;s|MAX_LAYOUTS *= *[0-9]\+|MAX_LAYOUTS = 300|&quot;</span> <span class="st">&quot;</span><span class="va">$strPath</span><span class="st">/server/common/constants.py&quot;</span></span>
<span id="cb11-30"><a href="methods.html#cb11-30" aria-hidden="true" tabindex="-1"></a>  <span class="cf">fi</span></span>
<span id="cb11-31"><a href="methods.html#cb11-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-32"><a href="methods.html#cb11-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">find</span> ./cellxgene/server/ <span class="at">-name</span> <span class="st">&quot;decode_fbs.py&quot;</span> <span class="at">-exec</span> cp {} <span class="va">$strPath</span>/server/app/. <span class="dt">\;</span></span>
<span id="cb11-33"><a href="methods.html#cb11-33" aria-hidden="true" tabindex="-1"></a><span class="cf">fi</span></span>
<span id="cb11-34"><a href="methods.html#cb11-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-35"><a href="methods.html#cb11-35" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="at">-e</span> <span class="st">&quot;\nls -l </span><span class="va">$strweb</span><span class="st">\n&quot;</span></span>
<span id="cb11-36"><a href="methods.html#cb11-36" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-l</span> <span class="va">$strweb</span></span></code></pre></div>
</div>
<div id="communication-between-vip-and-cellxgene-web-gui" class="section level2 hasAnchor" number="3.3">
<h2><span class="header-section-number">3.3</span> Communication between VIP and cellxgene web GUI<a href="methods.html#communication-between-vip-and-cellxgene-web-gui" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Cellxgene client utilizes React Redux that is the official React binding for Redux. It lets your React components read data from a Redux store, and dispatch actions to the store to update data.</p>
<p>So, this following code in config.sh appends “window.store = store;” to the end of client/src/reducers/index.js of cellxgene source code to expose the store to the browser.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="methods.html#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="at">-e</span> <span class="st">&quot;\nwindow.store = store;&quot;</span> <span class="op">&gt;&gt;</span> cellxgene/client/src/reducers/index.js</span></code></pre></div>
<p>By doing this, Redux store holding client data and user selections are visible to VIP to access variables and dispatch actions to control cellxgene user interface. For example,</p>
<ul>
<li>Unselect / select a feature. GUI is refreshed automatically after dispatching.</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="sourceCode js"><code class="sourceCode javascript"><span id="cb13-1"><a href="methods.html#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="bu">window</span><span class="op">.</span><span class="at">store</span><span class="op">.</span><span class="fu">dispatch</span>({<span class="dt">type</span><span class="op">:</span> <span class="st">&quot;categorical metadata filter deselect&quot;</span><span class="op">,</span> <span class="dt">metadataField</span><span class="op">:</span> <span class="st">&quot;louvain&quot;</span><span class="op">,</span> <span class="dt">categoryIndex</span><span class="op">:</span> <span class="dv">5</span>})</span>
<span id="cb13-2"><a href="methods.html#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="bu">window</span><span class="op">.</span><span class="at">store</span><span class="op">.</span><span class="fu">dispatch</span>({<span class="dt">type</span><span class="op">:</span> <span class="st">&quot;categorical metadata filter select&quot;</span><span class="op">,</span> <span class="dt">metadataField</span><span class="op">:</span> <span class="st">&quot;louvain&quot;</span><span class="op">,</span> <span class="dt">categoryIndex</span><span class="op">:</span> <span class="dv">5</span>})</span></code></pre></div>
<ul>
<li>Get the state of a just finished action and synchronize gene input and cell selections from main window to VIP if corresponding action was performed.</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="sourceCode js"><code class="sourceCode javascript"><span id="cb14-1"><a href="methods.html#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="kw">const</span> unsubscribe <span class="op">=</span> <span class="bu">window</span><span class="op">.</span><span class="at">store</span><span class="op">.</span><span class="fu">subscribe</span>(() <span class="kw">=&gt;</span> {</span>
<span id="cb14-2"><a href="methods.html#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="bu">window</span><span class="op">.</span><span class="at">store</span><span class="op">.</span><span class="fu">getState</span>()[<span class="st">&quot;@@undoable/filterState&quot;</span>]<span class="op">.</span><span class="at">prevAction</span>) {</span>
<span id="cb14-3"><a href="methods.html#cb14-3" aria-hidden="true" tabindex="-1"></a>    actionType <span class="op">=</span> <span class="bu">window</span><span class="op">.</span><span class="at">store</span><span class="op">.</span><span class="fu">getState</span>()[<span class="st">&quot;@@undoable/filterState&quot;</span>]<span class="op">.</span><span class="at">prevAction</span><span class="op">.</span><span class="at">type</span><span class="op">;</span></span>
<span id="cb14-4"><a href="methods.html#cb14-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (actionType<span class="op">.</span><span class="fu">includes</span>(<span class="st">&quot;user defined gene success&quot;</span>) <span class="op">||</span></span>
<span id="cb14-5"><a href="methods.html#cb14-5" aria-hidden="true" tabindex="-1"></a>    actionType<span class="op">.</span><span class="fu">includes</span>(<span class="st">&quot;store current cell selection as differential set&quot;</span>)) {</span>
<span id="cb14-6"><a href="methods.html#cb14-6" aria-hidden="true" tabindex="-1"></a>      <span class="fu">sync</span>()<span class="op">;</span></span>
<span id="cb14-7"><a href="methods.html#cb14-7" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb14-8"><a href="methods.html#cb14-8" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb14-9"><a href="methods.html#cb14-9" aria-hidden="true" tabindex="-1"></a>})<span class="op">;</span></span></code></pre></div>
</div>
<div id="diffxpy-integration" class="section level2 hasAnchor" number="3.4">
<h2><span class="header-section-number">3.4</span> Diffxpy Integration<a href="methods.html#diffxpy-integration" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This is the sample pseudocode, please see VIPInterface.py for actual implementation.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb15-1"><a href="methods.html#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb15-2"><a href="methods.html#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb15-3"><a href="methods.html#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> diffxpy.api <span class="im">as</span> app</span>
<span id="cb15-4"><a href="methods.html#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co"># set 1 of cells as cell1; set 2 of cells as cell2</span></span>
<span id="cb15-5"><a href="methods.html#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="methods.html#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="methods.html#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> app.get_data_adaptor() <span class="im">as</span> data_adaptor:</span>
<span id="cb15-8"><a href="methods.html#cb15-8" aria-hidden="true" tabindex="-1"></a>  X1 <span class="op">=</span> data_adaptor.data.X[cell1]</span>
<span id="cb15-9"><a href="methods.html#cb15-9" aria-hidden="true" tabindex="-1"></a>  X2 <span class="op">=</span> data_adaptor.data.X[cell2]</span>
<span id="cb15-10"><a href="methods.html#cb15-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-11"><a href="methods.html#cb15-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-12"><a href="methods.html#cb15-12" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.AnnData(pd.concat([X1,X2]),pd.DataFrame([<span class="st">&#39;grp1&#39;</span> <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(X1.shape[<span class="dv">0</span>])]<span class="op">+</span>[<span class="st">&#39;grp2&#39;</span> <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(X2.shape[<span class="dv">0</span>])],columns<span class="op">=</span>[<span class="st">&#39;comGrp&#39;</span>]))</span>
<span id="cb15-13"><a href="methods.html#cb15-13" aria-hidden="true" tabindex="-1"></a>deg <span class="op">=</span> de.test.two_sample(adata,<span class="st">&#39;comGrp&#39;</span>).summary()</span>
<span id="cb15-14"><a href="methods.html#cb15-14" aria-hidden="true" tabindex="-1"></a><span class="co">#deg is a dataframe contains the folloing columns [&#39;gene&#39;,&#39;log2fc&#39;,&#39;pval&#39;,&#39;qval&#39;]</span></span></code></pre></div>
</div>
<div id="create-a-h5ad-file-from-seurat-object" class="section level2 hasAnchor" number="3.5">
<h2><span class="header-section-number">3.5</span> Create a h5ad file from Seurat object<a href="methods.html#create-a-h5ad-file-from-seurat-object" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>First, export the following from Seurat object in R: <strong>expression matrix (assume normalized), metadata and coordinates (pca, tsne, umap) as separate txt files.</strong></p>
<p>Next in Python, create an AnnData object from 10x (scanpy.read_h5ad function) as a starting point. Then replace the expression matrix, meta data and coordinates as shown in the following Python code block to generate a h5ad file.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb16-1"><a href="methods.html#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb16-2"><a href="methods.html#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb16-3"><a href="methods.html#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb16-4"><a href="methods.html#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb16-5"><a href="methods.html#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb16-6"><a href="methods.html#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> numpy <span class="im">import</span> ndarray, unique</span>
<span id="cb16-7"><a href="methods.html#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.sparse.csc <span class="im">import</span> csc_matrix</span>
<span id="cb16-8"><a href="methods.html#cb16-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-9"><a href="methods.html#cb16-9" aria-hidden="true" tabindex="-1"></a>adata<span class="op">=</span> sc.read_h5ad(<span class="st">&quot;previous generated .h5ad&quot;</span>)</span>
<span id="cb16-10"><a href="methods.html#cb16-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-11"><a href="methods.html#cb16-11" aria-hidden="true" tabindex="-1"></a><span class="co"># read clustering res</span></span>
<span id="cb16-12"><a href="methods.html#cb16-12" aria-hidden="true" tabindex="-1"></a>xpca <span class="op">=</span> pd.read_csv(“.<span class="op">/</span>data<span class="op">/</span>harmony_clustered.h5ad.pca_coordinates.txt<span class="st">&quot;, sep=&#39;</span><span class="ch">\t</span><span class="st">&#39;, encoding=&#39;utf-8&#39;)</span></span>
<span id="cb16-13"><a href="methods.html#cb16-13" aria-hidden="true" tabindex="-1"></a><span class="er">xtsne = pd.read_csv(“./data/harmony_clustered.h5ad.tsne_coordinates.txt&quot;, sep=&#39;\t&#39;, encoding=&#39;utf-8&#39;)</span></span>
<span id="cb16-14"><a href="methods.html#cb16-14" aria-hidden="true" tabindex="-1"></a><span class="er">xumap = pd.read_csv(“./data/harmony_clustered.h5ad.umap_coordinates.txt&quot;, sep=&#39;\t&#39;, encoding=&#39;utf-8&#39;)</span></span>
<span id="cb16-15"><a href="methods.html#cb16-15" aria-hidden="true" tabindex="-1"></a><span class="er">xobs = pd.read_csv(“./data/harmony_clustered.h5ad.meta_data.txt&quot;, sep=&#39;\t&#39;, encoding=&#39;utf-8&#39;)</span></span>
<span id="cb16-16"><a href="methods.html#cb16-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-17"><a href="methods.html#cb16-17" aria-hidden="true" tabindex="-1"></a><span class="er">xpca.set_index(&#39;index&#39;, inplace=True)</span></span>
<span id="cb16-18"><a href="methods.html#cb16-18" aria-hidden="true" tabindex="-1"></a><span class="er">xtsne.set_index(&#39;index&#39;, inplace=True)</span></span>
<span id="cb16-19"><a href="methods.html#cb16-19" aria-hidden="true" tabindex="-1"></a><span class="er">xumap.set_index(&#39;index&#39;, inplace=True)</span></span>
<span id="cb16-20"><a href="methods.html#cb16-20" aria-hidden="true" tabindex="-1"></a><span class="er">xobs.set_index(&#39;index&#39;, inplace=True)</span></span>
<span id="cb16-21"><a href="methods.html#cb16-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-22"><a href="methods.html#cb16-22" aria-hidden="true" tabindex="-1"></a><span class="er">adata.obsm[&#39;X_pca&#39;] = np.array(xpca.loc[adataRaw.obs.index])</span></span>
<span id="cb16-23"><a href="methods.html#cb16-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-24"><a href="methods.html#cb16-24" aria-hidden="true" tabindex="-1"></a><span class="er">adata.obsm[&#39;X_tsne&#39;] = np.array(xtsne.loc[adataRaw.obs.index])</span></span>
<span id="cb16-25"><a href="methods.html#cb16-25" aria-hidden="true" tabindex="-1"></a><span class="er">adata.obsm[&#39;X_umap&#39;] = np.array(xumap.loc[adataRaw.obs.index])</span></span>
<span id="cb16-26"><a href="methods.html#cb16-26" aria-hidden="true" tabindex="-1"></a><span class="er">adata.obs = xobs.loc[adataRaw.obs.index] # this is a pandas dataframe</span></span>
<span id="cb16-27"><a href="methods.html#cb16-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-28"><a href="methods.html#cb16-28" aria-hidden="true" tabindex="-1"></a><span class="er"># read in expression matrix as numpy.ndarray</span></span>
<span id="cb16-29"><a href="methods.html#cb16-29" aria-hidden="true" tabindex="-1"></a><span class="er">exp_mat = np.loadtxt(fname =”expression matrix .txt&quot;)</span></span>
<span id="cb16-30"><a href="methods.html#cb16-30" aria-hidden="true" tabindex="-1"></a><span class="er">adata.X = exp_mat</span></span>
<span id="cb16-31"><a href="methods.html#cb16-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-32"><a href="methods.html#cb16-32" aria-hidden="true" tabindex="-1"></a><span class="er"># convert dense matrix into sparse matrix to save storage space and memory usage</span></span>
<span id="cb16-33"><a href="methods.html#cb16-33" aria-hidden="true" tabindex="-1"></a><span class="er">adata.X = csc_matrix(adata.X)_matrix</span></span>
<span id="cb16-34"><a href="methods.html#cb16-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-35"><a href="methods.html#cb16-35" aria-hidden="true" tabindex="-1"></a><span class="er"># add short description and initial graph settings. “|” and “by” are delimiters for VIP to parse the initial settings. Please follow the same rule for your own h5ad files.</span></span>
<span id="cb16-36"><a href="methods.html#cb16-36" aria-hidden="true" tabindex="-1"></a><span class="er">adata.obs[&#39;&gt;Description&#39;] = [&#39;Human brain snRNAseq 46k cells (MS Nature 2019 Schirmer et al.); data normalized, log transformed and scaled UMI; platform - 10X v2 chemistry | embedding by umap; color by cell_type&#39;]*adata.n_obs</span></span>
<span id="cb16-37"><a href="methods.html#cb16-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-38"><a href="methods.html#cb16-38" aria-hidden="true" tabindex="-1"></a><span class="er"># Then last step to save h5ad:</span></span>
<span id="cb16-39"><a href="methods.html#cb16-39" aria-hidden="true" tabindex="-1"></a><span class="er">adata.write_h5ad(&quot;final output.h5ad&quot;)</span></span></code></pre></div>
<p>When the h5ad file is uploaded to cellxgeneVIP, AnnData.X matrix is to be used for visualization and DEG analysis. By default, the data (e.g, raw count matrix) is assumed to be unscaled , however, if the data have been scaled or normalized, the user needs to turn off the option ‘Scale data to unit variance for plotting:’ in ‘Global Setting’.</p>
</div>
<div id="prepare-spatial-data-for-visualization" class="section level2 hasAnchor" number="3.6">
<h2><span class="header-section-number">3.6</span> Prepare Spatial Data for Visualization<a href="methods.html#prepare-spatial-data-for-visualization" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="x-visium-data" class="section level3 hasAnchor" number="3.6.1">
<h3><span class="header-section-number">3.6.1</span> 10X visium data<a href="methods.html#x-visium-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This is the sample script for the spatial demo data set adapted from <a href="https://github.com/interactivereport/cellxgene_VIP/blob/master/notebook/spatial_transcriptomics.ipynb">cellxgene_VIP spatial transcriptomics notebook</a>. Please go to <a href="https://github.com/interactivereport/cellxgene_VIP/blob/master/notebook/spatial_transcriptomics.ipynb">cellxgene_VIP spatial transcriptomics notebook</a> to check the intermediate outputs.</p>
<div id="download-input-demo-data" class="section level4 hasAnchor" number="3.6.1.1">
<h4><span class="header-section-number">3.6.1.1</span> Download input demo data<a href="methods.html#download-input-demo-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb17-1"><a href="methods.html#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-O</span> 10X_ST_demo.tar.gz https://zenodo.org/record/5765589/files/10X_ST_demo.tar.gz<span class="pp">?</span>download=1</span>
<span id="cb17-2"><a href="methods.html#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="methods.html#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-zxvf</span> 10X_ST_demo.tar.gz</span></code></pre></div>
</div>
<div id="using-python-to-prepare-10x-visium-spatial-data" class="section level4 hasAnchor" number="3.6.1.2">
<h4><span class="header-section-number">3.6.1.2</span> Using python to prepare 10X visium spatial data<a href="methods.html#using-python-to-prepare-10x-visium-spatial-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb18"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb18-1"><a href="methods.html#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb18-2"><a href="methods.html#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb18-3"><a href="methods.html#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys, getopt</span>
<span id="cb18-4"><a href="methods.html#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb18-5"><a href="methods.html#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb18-6"><a href="methods.html#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> copy</span>
<span id="cb18-7"><a href="methods.html#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb18-8"><a href="methods.html#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> PIL <span class="im">import</span> Image, ImageOps</span>
<span id="cb18-9"><a href="methods.html#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> </span>
<span id="cb18-10"><a href="methods.html#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata</span>
<span id="cb18-11"><a href="methods.html#cb18-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-12"><a href="methods.html#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="co"># change directory to 10X_ST_demo which is just downloaded, if in current directory</span></span>
<span id="cb18-13"><a href="methods.html#cb18-13" aria-hidden="true" tabindex="-1"></a>os.chdir(<span class="st">&#39;./10X_ST_demo&#39;</span>)</span>
<span id="cb18-14"><a href="methods.html#cb18-14" aria-hidden="true" tabindex="-1"></a>inputfile<span class="op">=</span><span class="st">&#39;inputfiles.txt&#39;</span></span>
<span id="cb18-15"><a href="methods.html#cb18-15" aria-hidden="true" tabindex="-1"></a>samples <span class="op">=</span> pd.read_csv(inputfile, header <span class="op">=</span> <span class="va">None</span>)[<span class="dv">0</span>]</span>
<span id="cb18-16"><a href="methods.html#cb18-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-17"><a href="methods.html#cb18-17" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> read_each(i):</span>
<span id="cb18-18"><a href="methods.html#cb18-18" aria-hidden="true" tabindex="-1"></a>    adata <span class="op">=</span> sc.read_visium(i)</span>
<span id="cb18-19"><a href="methods.html#cb18-19" aria-hidden="true" tabindex="-1"></a>    adata.var_names_make_unique()</span>
<span id="cb18-20"><a href="methods.html#cb18-20" aria-hidden="true" tabindex="-1"></a>    <span class="co"># flip Y axis to show correctly in cellxgene VIP</span></span>
<span id="cb18-21"><a href="methods.html#cb18-21" aria-hidden="true" tabindex="-1"></a>    adata.obsm[<span class="st">&#39;spatial&#39;</span>][:,<span class="dv">1</span>] <span class="op">=</span> <span class="op">-</span>adata.obsm[<span class="st">&#39;spatial&#39;</span>][:,<span class="dv">1</span>]</span>
<span id="cb18-22"><a href="methods.html#cb18-22" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span>(adata)</span>
<span id="cb18-23"><a href="methods.html#cb18-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-24"><a href="methods.html#cb18-24" aria-hidden="true" tabindex="-1"></a>adatals <span class="op">=</span> [read_each(i) <span class="cf">for</span> i <span class="kw">in</span> samples]</span>
<span id="cb18-25"><a href="methods.html#cb18-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-26"><a href="methods.html#cb18-26" aria-hidden="true" tabindex="-1"></a>sampleIDs <span class="op">=</span> samples.<span class="bu">str</span>.extract(<span class="vs">r&#39;10X_demo_data_(.*)&#39;</span>)</span>
<span id="cb18-27"><a href="methods.html#cb18-27" aria-hidden="true" tabindex="-1"></a>sampleIDs <span class="op">=</span> <span class="st">&quot;V1_&quot;</span><span class="op">+</span> sampleIDs</span>
<span id="cb18-28"><a href="methods.html#cb18-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-29"><a href="methods.html#cb18-29" aria-hidden="true" tabindex="-1"></a>adata_merge <span class="op">=</span> sc.AnnData.concatenate(<span class="op">*</span>adatals, batch_key<span class="op">=</span><span class="st">&#39;sample&#39;</span>, join<span class="op">=</span><span class="st">&#39;outer&#39;</span>, batch_categories<span class="op">=</span> sampleIDs[<span class="dv">0</span>].astype(<span class="st">&quot;category&quot;</span>))</span>
<span id="cb18-30"><a href="methods.html#cb18-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-31"><a href="methods.html#cb18-31" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">list</span>(<span class="bu">range</span>(<span class="bu">len</span>(adatals))):</span>
<span id="cb18-32"><a href="methods.html#cb18-32" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(i)</span>
<span id="cb18-33"><a href="methods.html#cb18-33" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add back the spatial coordinates as separate embeddings</span></span>
<span id="cb18-34"><a href="methods.html#cb18-34" aria-hidden="true" tabindex="-1"></a>    adata_merge.obsm[<span class="st">&#39;X_spatial_&#39;</span><span class="op">+</span><span class="bu">list</span>(adatals[i].uns[<span class="st">&quot;spatial&quot;</span>])[<span class="dv">0</span>]] <span class="op">=</span> np.zeros(adata_merge.obsm[<span class="st">&#39;spatial&#39;</span>].shape)</span>
<span id="cb18-35"><a href="methods.html#cb18-35" aria-hidden="true" tabindex="-1"></a>    adata_merge.obsm[<span class="st">&#39;X_spatial_&#39;</span><span class="op">+</span><span class="bu">list</span>(adatals[i].uns[<span class="st">&quot;spatial&quot;</span>])[<span class="dv">0</span>]][np.where(adata_merge.obs[<span class="st">&#39;sample&#39;</span>]<span class="op">==</span><span class="bu">list</span>(adatals[i].uns[<span class="st">&quot;spatial&quot;</span>])[<span class="dv">0</span>])] <span class="op">=</span> adatals[i].obsm[<span class="st">&#39;spatial&#39;</span>]</span>
<span id="cb18-36"><a href="methods.html#cb18-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-37"><a href="methods.html#cb18-37" aria-hidden="true" tabindex="-1"></a>adata_merge.uns[<span class="st">&#39;spatial&#39;</span>] <span class="op">=</span> <span class="bu">dict</span>()</span>
<span id="cb18-38"><a href="methods.html#cb18-38" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">list</span>(<span class="bu">range</span>(<span class="bu">len</span>(adatals))):</span>
<span id="cb18-39"><a href="methods.html#cb18-39" aria-hidden="true" tabindex="-1"></a>    adata_merge.uns[<span class="st">&#39;spatial&#39;</span>][<span class="st">&quot;spatial_&quot;</span><span class="op">+</span><span class="bu">list</span>(adatals[i].uns[<span class="st">&quot;spatial&quot;</span>])[<span class="dv">0</span>]] <span class="op">=</span> adatals[i].uns[<span class="st">&#39;spatial&#39;</span>][<span class="bu">list</span>(adatals[i].uns[<span class="st">&quot;spatial&quot;</span>])[<span class="dv">0</span>]]</span>
<span id="cb18-40"><a href="methods.html#cb18-40" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb18-41"><a href="methods.html#cb18-41" aria-hidden="true" tabindex="-1"></a><span class="co"># QC metric</span></span>
<span id="cb18-42"><a href="methods.html#cb18-42" aria-hidden="true" tabindex="-1"></a>adata_merge.var[<span class="st">&quot;mt&quot;</span>] <span class="op">=</span> adata_merge.var_names.<span class="bu">str</span>.startswith(<span class="st">&quot;MT-&quot;</span>)</span>
<span id="cb18-43"><a href="methods.html#cb18-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-44"><a href="methods.html#cb18-44" aria-hidden="true" tabindex="-1"></a>sc.pp.calculate_qc_metrics(adata_merge, qc_vars<span class="op">=</span>[<span class="st">&quot;mt&quot;</span>], inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb18-45"><a href="methods.html#cb18-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-46"><a href="methods.html#cb18-46" aria-hidden="true" tabindex="-1"></a><span class="co"># QC plots</span></span>
<span id="cb18-47"><a href="methods.html#cb18-47" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">4</span>, figsize<span class="op">=</span>(<span class="dv">15</span>, <span class="dv">4</span>))</span>
<span id="cb18-48"><a href="methods.html#cb18-48" aria-hidden="true" tabindex="-1"></a>sns.distplot(adata_merge.obs[<span class="st">&quot;total_counts&quot;</span>], kde<span class="op">=</span><span class="va">False</span>, ax<span class="op">=</span>axs[<span class="dv">0</span>])</span>
<span id="cb18-49"><a href="methods.html#cb18-49" aria-hidden="true" tabindex="-1"></a>sns.distplot(adata_merge.obs[<span class="st">&quot;total_counts&quot;</span>][adata_merge.obs[<span class="st">&quot;total_counts&quot;</span>] <span class="op">&lt;</span> <span class="dv">10000</span>], kde<span class="op">=</span><span class="va">False</span>, bins<span class="op">=</span><span class="dv">40</span>, ax<span class="op">=</span>axs[<span class="dv">1</span>])</span>
<span id="cb18-50"><a href="methods.html#cb18-50" aria-hidden="true" tabindex="-1"></a>sns.distplot(adata_merge.obs[<span class="st">&quot;n_genes_by_counts&quot;</span>], kde<span class="op">=</span><span class="va">False</span>, bins<span class="op">=</span><span class="dv">60</span>, ax<span class="op">=</span>axs[<span class="dv">2</span>])</span>
<span id="cb18-51"><a href="methods.html#cb18-51" aria-hidden="true" tabindex="-1"></a>sns.distplot(adata_merge.obs[<span class="st">&quot;n_genes_by_counts&quot;</span>][adata_merge.obs[<span class="st">&quot;n_genes_by_counts&quot;</span>] <span class="op">&lt;</span> <span class="dv">4000</span>], kde<span class="op">=</span><span class="va">False</span>, bins<span class="op">=</span><span class="dv">60</span>, ax<span class="op">=</span>axs[<span class="dv">3</span>])</span>
<span id="cb18-52"><a href="methods.html#cb18-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-53"><a href="methods.html#cb18-53" aria-hidden="true" tabindex="-1"></a><span class="co"># filtering, turn this off to keep all spots for visualization also cutoffs are case-by-case based on the QC plots</span></span>
<span id="cb18-54"><a href="methods.html#cb18-54" aria-hidden="true" tabindex="-1"></a><span class="co">#sc.pp.filter_cells(adata, min_counts=5000)</span></span>
<span id="cb18-55"><a href="methods.html#cb18-55" aria-hidden="true" tabindex="-1"></a><span class="co">#sc.pp.filter_cells(adata, max_counts=35000)</span></span>
<span id="cb18-56"><a href="methods.html#cb18-56" aria-hidden="true" tabindex="-1"></a><span class="co">#adata = adata[adata.obs[&quot;pct_counts_mt&quot;] &lt; 20]</span></span>
<span id="cb18-57"><a href="methods.html#cb18-57" aria-hidden="true" tabindex="-1"></a><span class="co">#print(f&quot;#cells after MT filter: {adata.n_obs}&quot;)</span></span>
<span id="cb18-58"><a href="methods.html#cb18-58" aria-hidden="true" tabindex="-1"></a><span class="co">#sc.pp.filter_genes(adata, min_cells=10)</span></span>
<span id="cb18-59"><a href="methods.html#cb18-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-60"><a href="methods.html#cb18-60" aria-hidden="true" tabindex="-1"></a><span class="co"># normalization, log1p transformation and select HVGs</span></span>
<span id="cb18-61"><a href="methods.html#cb18-61" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_total(adata_merge, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb18-62"><a href="methods.html#cb18-62" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(adata_merge)</span>
<span id="cb18-63"><a href="methods.html#cb18-63" aria-hidden="true" tabindex="-1"></a>sc.pp.highly_variable_genes(adata_merge, flavor<span class="op">=</span><span class="st">&quot;seurat&quot;</span>, n_top_genes<span class="op">=</span><span class="dv">2000</span>)</span>
<span id="cb18-64"><a href="methods.html#cb18-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-65"><a href="methods.html#cb18-65" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA, UMAP and clustering by leiden</span></span>
<span id="cb18-66"><a href="methods.html#cb18-66" aria-hidden="true" tabindex="-1"></a>sc.pp.pca(adata_merge)</span>
<span id="cb18-67"><a href="methods.html#cb18-67" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata_merge)</span>
<span id="cb18-68"><a href="methods.html#cb18-68" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(adata_merge)</span>
<span id="cb18-69"><a href="methods.html#cb18-69" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata_merge, key_added<span class="op">=</span><span class="st">&quot;clusters&quot;</span>)</span>
<span id="cb18-70"><a href="methods.html#cb18-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-71"><a href="methods.html#cb18-71" aria-hidden="true" tabindex="-1"></a><span class="co"># collect sample names</span></span>
<span id="cb18-72"><a href="methods.html#cb18-72" aria-hidden="true" tabindex="-1"></a>sampleNames <span class="op">=</span> <span class="bu">list</span>()</span>
<span id="cb18-73"><a href="methods.html#cb18-73" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> f <span class="kw">in</span> <span class="bu">list</span>(adata_merge.obsm):</span>
<span id="cb18-74"><a href="methods.html#cb18-74" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="st">&quot;spatial_&quot;</span> <span class="kw">in</span> f: <span class="co"># search for the pattern</span></span>
<span id="cb18-75"><a href="methods.html#cb18-75" aria-hidden="true" tabindex="-1"></a>        library_id<span class="op">=</span>f.replace(<span class="st">&quot;X_spatial_&quot;</span>,<span class="st">&quot;&quot;</span>) <span class="co"># parse the string and get the sample id</span></span>
<span id="cb18-76"><a href="methods.html#cb18-76" aria-hidden="true" tabindex="-1"></a>        <span class="co">#library_id=library_id.replace(&quot;V1_&quot;,&quot;&quot;)</span></span>
<span id="cb18-77"><a href="methods.html#cb18-77" aria-hidden="true" tabindex="-1"></a>        sampleNames.append(library_id)</span>
<span id="cb18-78"><a href="methods.html#cb18-78" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb18-79"><a href="methods.html#cb18-79" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> PIL <span class="im">import</span> Image</span>
<span id="cb18-80"><a href="methods.html#cb18-80" aria-hidden="true" tabindex="-1"></a>spatial<span class="op">=</span>adata_merge.uns[<span class="st">&quot;spatial&quot;</span>]</span>
<span id="cb18-81"><a href="methods.html#cb18-81" aria-hidden="true" tabindex="-1"></a>dim<span class="op">=</span><span class="st">&#39;&#39;</span></span>
<span id="cb18-82"><a href="methods.html#cb18-82" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> math</span>
<span id="cb18-83"><a href="methods.html#cb18-83" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> dim<span class="op">==</span><span class="st">&#39;&#39;</span>:</span>
<span id="cb18-84"><a href="methods.html#cb18-84" aria-hidden="true" tabindex="-1"></a>    height <span class="op">=</span> math.ceil(math.sqrt(<span class="bu">len</span>(samples)))</span>
<span id="cb18-85"><a href="methods.html#cb18-85" aria-hidden="true" tabindex="-1"></a>    width <span class="op">=</span> math.ceil(<span class="bu">len</span>(samples)<span class="op">/</span>height)</span>
<span id="cb18-86"><a href="methods.html#cb18-86" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:</span>
<span id="cb18-87"><a href="methods.html#cb18-87" aria-hidden="true" tabindex="-1"></a>    width,height <span class="op">=</span> dim.split(<span class="st">&#39;x&#39;</span>)</span>
<span id="cb18-88"><a href="methods.html#cb18-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-89"><a href="methods.html#cb18-89" aria-hidden="true" tabindex="-1"></a>idx <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb18-90"><a href="methods.html#cb18-90" aria-hidden="true" tabindex="-1"></a>size<span class="op">=</span><span class="dv">700</span></span>
<span id="cb18-91"><a href="methods.html#cb18-91" aria-hidden="true" tabindex="-1"></a><span class="co">#creates a new empty image, RGB mode, and size 1400 by 1400.</span></span>
<span id="cb18-92"><a href="methods.html#cb18-92" aria-hidden="true" tabindex="-1"></a>new_im <span class="op">=</span> Image.new(<span class="st">&#39;RGB&#39;</span>, (size<span class="op">*</span>width,size<span class="op">*</span>height))</span>
<span id="cb18-93"><a href="methods.html#cb18-93" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,size<span class="op">*</span>width,size):</span>
<span id="cb18-94"><a href="methods.html#cb18-94" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> j <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,size<span class="op">*</span>height,size):</span>
<span id="cb18-95"><a href="methods.html#cb18-95" aria-hidden="true" tabindex="-1"></a>        <span class="co"># load the image from the object</span></span>
<span id="cb18-96"><a href="methods.html#cb18-96" aria-hidden="true" tabindex="-1"></a>        <span class="co">#im = Image.fromarray((spatial[&quot;spatial_V1_&quot;+samples[idx]][&quot;images&quot;][&quot;lowres&quot;]* 255).round().astype(np.uint8)) # found a solution to covert float32 to unit8</span></span>
<span id="cb18-97"><a href="methods.html#cb18-97" aria-hidden="true" tabindex="-1"></a>        im <span class="op">=</span> Image.fromarray((spatial[<span class="st">&quot;spatial_&quot;</span><span class="op">+</span>sampleNames[idx]][<span class="st">&quot;images&quot;</span>][<span class="st">&quot;lowres&quot;</span>]<span class="op">*</span> <span class="dv">255</span>).<span class="bu">round</span>().astype(np.uint8)) <span class="co"># found a solution to covert float32 to unit8</span></span>
<span id="cb18-98"><a href="methods.html#cb18-98" aria-hidden="true" tabindex="-1"></a>        <span class="co"># paste images together</span></span>
<span id="cb18-99"><a href="methods.html#cb18-99" aria-hidden="true" tabindex="-1"></a>        new_im.paste(im, (j,i))</span>
<span id="cb18-100"><a href="methods.html#cb18-100" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(idx)</span>
<span id="cb18-101"><a href="methods.html#cb18-101" aria-hidden="true" tabindex="-1"></a>        idx <span class="op">=</span> idx<span class="op">+</span><span class="dv">1</span></span>
<span id="cb18-102"><a href="methods.html#cb18-102" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> idx<span class="op">&gt;=</span><span class="bu">len</span>(sampleNames):</span>
<span id="cb18-103"><a href="methods.html#cb18-103" aria-hidden="true" tabindex="-1"></a>            <span class="cf">break</span></span>
<span id="cb18-104"><a href="methods.html#cb18-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-105"><a href="methods.html#cb18-105" aria-hidden="true" tabindex="-1"></a><span class="co"># fake a adata.uns by providing merged lowres image and scale factors 1</span></span>
<span id="cb18-106"><a href="methods.html#cb18-106" aria-hidden="true" tabindex="-1"></a>adata_merge.uns[<span class="st">&#39;spatial&#39;</span>][<span class="st">&#39;spatial_Merged&#39;</span>] <span class="op">=</span> copy.deepcopy(adata_merge.uns[<span class="st">&#39;spatial&#39;</span>][<span class="bu">list</span>(adata_merge.uns[<span class="st">&#39;spatial&#39;</span>])[<span class="dv">0</span>]])</span>
<span id="cb18-107"><a href="methods.html#cb18-107" aria-hidden="true" tabindex="-1"></a>adata_merge.uns[<span class="st">&#39;spatial&#39;</span>][<span class="st">&#39;spatial_Merged&#39;</span>][<span class="st">&#39;images&#39;</span>][<span class="st">&quot;hires&quot;</span>] <span class="op">=</span> np.asarray(new_im)</span>
<span id="cb18-108"><a href="methods.html#cb18-108" aria-hidden="true" tabindex="-1"></a>adata_merge.uns[<span class="st">&#39;spatial&#39;</span>][<span class="st">&#39;spatial_Merged&#39;</span>][<span class="st">&#39;images&#39;</span>][<span class="st">&quot;lowres&quot;</span>] <span class="op">=</span> np.asarray(new_im)</span>
<span id="cb18-109"><a href="methods.html#cb18-109" aria-hidden="true" tabindex="-1"></a>adata_merge.uns[<span class="st">&#39;spatial&#39;</span>][<span class="st">&#39;spatial_Merged&#39;</span>][<span class="st">&#39;scalefactors&#39;</span>][<span class="st">&#39;tissue_lowres_scalef&#39;</span>] <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb18-110"><a href="methods.html#cb18-110" aria-hidden="true" tabindex="-1"></a>adata_merge.uns[<span class="st">&#39;spatial&#39;</span>][<span class="st">&#39;spatial_Merged&#39;</span>][<span class="st">&#39;scalefactors&#39;</span>][<span class="st">&#39;tissue_hires_scalef&#39;</span>] <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb18-111"><a href="methods.html#cb18-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-112"><a href="methods.html#cb18-112" aria-hidden="true" tabindex="-1"></a><span class="co"># add back the spatial coordinates as separate embeddings</span></span>
<span id="cb18-113"><a href="methods.html#cb18-113" aria-hidden="true" tabindex="-1"></a>idx <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb18-114"><a href="methods.html#cb18-114" aria-hidden="true" tabindex="-1"></a>adata_merge.obsm[<span class="st">&#39;X_spatial_Merged&#39;</span>] <span class="op">=</span> adata_merge.obsm[<span class="st">&#39;spatial&#39;</span>]</span>
<span id="cb18-115"><a href="methods.html#cb18-115" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,size<span class="op">*</span>width,size):</span>
<span id="cb18-116"><a href="methods.html#cb18-116" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> j <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,size<span class="op">*</span>height,size):</span>
<span id="cb18-117"><a href="methods.html#cb18-117" aria-hidden="true" tabindex="-1"></a>        <span class="co">#library_id=&#39;spatial_V1_&#39;+samples[idx] # parse the string and get the sample id</span></span>
<span id="cb18-118"><a href="methods.html#cb18-118" aria-hidden="true" tabindex="-1"></a>        library_id<span class="op">=</span><span class="st">&#39;spatial_&#39;</span><span class="op">+</span>sampleNames[idx] <span class="co"># parse the string and get the sample id</span></span>
<span id="cb18-119"><a href="methods.html#cb18-119" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(library_id)</span>
<span id="cb18-120"><a href="methods.html#cb18-120" aria-hidden="true" tabindex="-1"></a>        tissue_lowres_scalef <span class="op">=</span> spatial[library_id][<span class="st">&#39;scalefactors&#39;</span>][<span class="st">&#39;tissue_lowres_scalef&#39;</span>]</span>
<span id="cb18-121"><a href="methods.html#cb18-121" aria-hidden="true" tabindex="-1"></a>        adata_merge.obsm[<span class="st">&#39;X_spatial_Merged&#39;</span>][np.where(adata_merge.obs[<span class="st">&#39;sample&#39;</span>]<span class="op">==</span>sampleNames[idx])] <span class="op">=</span> copy.deepcopy(adatals[idx].obsm[<span class="st">&#39;spatial&#39;</span>])</span>
<span id="cb18-122"><a href="methods.html#cb18-122" aria-hidden="true" tabindex="-1"></a>        adata_merge.obsm[<span class="st">&#39;X_spatial_Merged&#39;</span>][np.where(adata_merge.obs[<span class="st">&#39;sample&#39;</span>]<span class="op">==</span>sampleNames[idx]),<span class="dv">1</span>] <span class="op">=</span> adatals[idx].obsm[<span class="st">&#39;spatial&#39;</span>][:,<span class="dv">1</span>]<span class="op">*</span>tissue_lowres_scalef <span class="op">-</span> i</span>
<span id="cb18-123"><a href="methods.html#cb18-123" aria-hidden="true" tabindex="-1"></a>        adata_merge.obsm[<span class="st">&#39;X_spatial_Merged&#39;</span>][np.where(adata_merge.obs[<span class="st">&#39;sample&#39;</span>]<span class="op">==</span>sampleNames[idx]),<span class="dv">0</span>] <span class="op">=</span> adatals[idx].obsm[<span class="st">&#39;spatial&#39;</span>][:,<span class="dv">0</span>]<span class="op">*</span>tissue_lowres_scalef <span class="op">+</span> j</span>
<span id="cb18-124"><a href="methods.html#cb18-124" aria-hidden="true" tabindex="-1"></a>        idx <span class="op">=</span> idx<span class="op">+</span><span class="dv">1</span></span>
<span id="cb18-125"><a href="methods.html#cb18-125" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> idx<span class="op">&gt;=</span><span class="bu">len</span>(sampleNames):</span>
<span id="cb18-126"><a href="methods.html#cb18-126" aria-hidden="true" tabindex="-1"></a>            <span class="cf">break</span></span>
<span id="cb18-127"><a href="methods.html#cb18-127" aria-hidden="true" tabindex="-1"></a>outputfile <span class="op">=</span> <span class="st">&#39;10X_data.h5ad&#39;</span></span>
<span id="cb18-128"><a href="methods.html#cb18-128" aria-hidden="true" tabindex="-1"></a>adata_merge.write_h5ad(outputfile)</span></code></pre></div>
</div>
</div>
<div id="nanostring-cosmx-data" class="section level3 hasAnchor" number="3.6.2">
<h3><span class="header-section-number">3.6.2</span> NanoString CosMx data<a href="methods.html#nanostring-cosmx-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This is the sample script for the spatial demo data set adapted from <a href="https://github.com/interactivereport/cellxgene_VIP/blob/master/notebook/public_cosMx_liver.ipynb">cellxgene_VIP public cosMx liver notebook</a>. Please go to <a href="https://github.com/interactivereport/cellxgene_VIP/blob/master/notebook/public_cosMx_liver.ipynb">cellxgene_VIP public cosMx liver notebook</a> to check the intermediate outputs.</p>
<div id="download-input-demo-data-1" class="section level4 hasAnchor" number="3.6.2.1">
<h4><span class="header-section-number">3.6.2.1</span> Download input demo data<a href="methods.html#download-input-demo-data-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Download the public cosMx data from <a href="https://nanostring.com/products/cosmx-spatial-molecular-imager/ffpe-dataset/">NanoString website</a>. The seurat object contains the expression, cell meta and transcript coordinates(some may contains cell boundaries). If the cell boundaries are not included in the seurat, they can be estimated from the transcript coordinates by cellPoly. The following will use <a href="https://nanostring.com/resources/seurat-object-cosmx-smi-human-liver-ffpe-dataset/">Human Liver (RNA)</a> as an example.</p>
</div>
<div id="export-expression-cell-meta-transcript-coordinate-and-reduction-umap-from-seurat-object-using-r" class="section level4 hasAnchor" number="3.6.2.2">
<h4><span class="header-section-number">3.6.2.2</span> Export Expression, cell meta, transcript coordinate and Reduction UMAP from seurat object using R<a href="methods.html#export-expression-cell-meta-transcript-coordinate-and-reduction-umap-from-seurat-object-using-r" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="methods.html#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(Seurat)</span>
<span id="cb19-2"><a href="methods.html#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(CellPoly)</span>
<span id="cb19-3"><a href="methods.html#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(dplyr)</span>
<span id="cb19-4"><a href="methods.html#cb19-4" aria-hidden="true" tabindex="-1"></a>D <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&#39;LiverDataReleaseSeurat_newUMAP.RDS&#39;</span>)</span>
<span id="cb19-5"><a href="methods.html#cb19-5" aria-hidden="true" tabindex="-1"></a>saveSlideFOV <span class="ot">&lt;-</span> <span class="cf">function</span>(slide,fov){</span>
<span id="cb19-6"><a href="methods.html#cb19-6" aria-hidden="true" tabindex="-1"></a>    selC <span class="ot">&lt;-</span> <span class="fu">rownames</span>(D<span class="sc">@</span>meta.data)[D<span class="sc">@</span>meta.data<span class="sc">$</span>fov<span class="sc">==</span>fov<span class="sc">&amp;</span>D<span class="sc">@</span>meta.data<span class="sc">$</span>slide_ID_numeric<span class="sc">==</span>slide]</span>
<span id="cb19-7"><a href="methods.html#cb19-7" aria-hidden="true" tabindex="-1"></a>    prefix <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;Slide&quot;</span>,slide,<span class="st">&quot;_FOV&quot;</span>,<span class="fu">gsub</span>(<span class="st">&quot; &quot;</span>,<span class="st">&quot;0&quot;</span>,<span class="fu">format</span>(fov,<span class="at">width=</span><span class="dv">3</span>)))</span>
<span id="cb19-8"><a href="methods.html#cb19-8" aria-hidden="true" tabindex="-1"></a>    data.table<span class="sc">::</span><span class="fu">fwrite</span>(<span class="fu">data.frame</span>(<span class="at">gene=</span><span class="fu">rownames</span>(D<span class="sc">@</span>assays<span class="sc">$</span>RNA),D<span class="sc">@</span>assays<span class="sc">$</span>RNA[,selC]),<span class="fu">paste0</span>(prefix,<span class="st">&#39;_exp.csv&#39;</span>))</span>
<span id="cb19-9"><a href="methods.html#cb19-9" aria-hidden="true" tabindex="-1"></a>    data.table<span class="sc">::</span><span class="fu">fwrite</span>(<span class="fu">data.frame</span>(<span class="at">cID=</span>selC,D<span class="sc">@</span>meta.data[selC,]),<span class="fu">paste0</span>(prefix,<span class="st">&#39;_meta.csv&#39;</span>))</span>
<span id="cb19-10"><a href="methods.html#cb19-10" aria-hidden="true" tabindex="-1"></a>    data.table<span class="sc">::</span><span class="fu">fwrite</span>(<span class="fu">data.frame</span>(<span class="at">cID=</span>selC,D<span class="sc">@</span>reductions<span class="sc">$</span>approximateumap<span class="sc">@</span>cell.embeddings[selC,]),<span class="fu">paste0</span>(prefix,<span class="st">&#39;_UMAP.csv&#39;</span>))</span>
<span id="cb19-11"><a href="methods.html#cb19-11" aria-hidden="true" tabindex="-1"></a>    data.table<span class="sc">::</span><span class="fu">fwrite</span>(<span class="fu">data.frame</span>(<span class="at">cID=</span>selC,D<span class="sc">@</span>reductions<span class="sc">$</span>approximateUMAP_bySlide<span class="sc">@</span>cell.embeddings[selC,]),<span class="fu">paste0</span>(prefix,<span class="st">&#39;_slideUMAP.csv&#39;</span>))</span>
<span id="cb19-12"><a href="methods.html#cb19-12" aria-hidden="true" tabindex="-1"></a>    data.table<span class="sc">::</span><span class="fu">fwrite</span>(D<span class="sc">@</span>misc<span class="sc">$</span>transcriptCoords[D<span class="sc">@</span>misc<span class="sc">$</span>transcriptCoords<span class="sc">$</span>fov<span class="sc">==</span>fov<span class="sc">&amp;</span>D<span class="sc">@</span>misc<span class="sc">$</span>transcriptCoords<span class="sc">$</span>slideID<span class="sc">==</span>slide,] <span class="sc">%&gt;%</span></span>
<span id="cb19-13"><a href="methods.html#cb19-13" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">cell_ID=</span>cell_id),<span class="fu">paste0</span>(prefix,<span class="st">&#39;_tx.csv&#39;</span>))</span>
<span id="cb19-14"><a href="methods.html#cb19-14" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb19-15"><a href="methods.html#cb19-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-16"><a href="methods.html#cb19-16" aria-hidden="true" tabindex="-1"></a><span class="fu">saveSlideFOV</span>(<span class="dv">1</span>,<span class="dv">2</span>)</span>
<span id="cb19-17"><a href="methods.html#cb19-17" aria-hidden="true" tabindex="-1"></a><span class="fu">saveSlideFOV</span>(<span class="dv">1</span>,<span class="dv">3</span>)</span></code></pre></div>
</div>
<div id="extract-cell-boundaries-from-cell-id-labeled-tif-file-using-python" class="section level4 hasAnchor" number="3.6.2.3">
<h4><span class="header-section-number">3.6.2.3</span> Extract cell boundaries from cell ID labeled tif file using python<a href="methods.html#extract-cell-boundaries-from-cell-id-labeled-tif-file-using-python" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb20"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb20-1"><a href="methods.html#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> cv2</span>
<span id="cb20-2"><a href="methods.html#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb20-3"><a href="methods.html#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb20-4"><a href="methods.html#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb20-5"><a href="methods.html#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> extractCellBoundaryPolygon(fov):</span>
<span id="cb20-6"><a href="methods.html#cb20-6" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(fov)</span>
<span id="cb20-7"><a href="methods.html#cb20-7" aria-hidden="true" tabindex="-1"></a>    img <span class="op">=</span> plt.imread((<span class="st">&#39;NormalLiverFiles/CellStatsDir/FOV</span><span class="sc">%*d</span><span class="st">/CellLabels_F</span><span class="sc">%*d</span><span class="st">.tif&#39;</span><span class="op">%</span>(<span class="dv">3</span>,fov,<span class="dv">3</span>,fov)).replace(<span class="st">&quot; &quot;</span>,<span class="st">&quot;0&quot;</span>))</span>
<span id="cb20-8"><a href="methods.html#cb20-8" aria-hidden="true" tabindex="-1"></a>    contours <span class="op">=</span> []</span>
<span id="cb20-9"><a href="methods.html#cb20-9" aria-hidden="true" tabindex="-1"></a>    nStep <span class="op">=</span> <span class="bu">int</span>(np.<span class="bu">max</span>(img)<span class="op">/</span><span class="dv">10</span>)</span>
<span id="cb20-10"><a href="methods.html#cb20-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>,np.<span class="bu">max</span>(img)<span class="op">+</span><span class="dv">1</span>):</span>
<span id="cb20-11"><a href="methods.html#cb20-11" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> i<span class="op">%</span>nStep<span class="op">==</span><span class="dv">0</span>:</span>
<span id="cb20-12"><a href="methods.html#cb20-12" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(<span class="st">&quot;</span><span class="sc">%d</span><span class="st">0</span><span class="sc">%%</span><span class="st">&quot;</span><span class="op">%</span>(i<span class="op">/</span>nStep),end<span class="op">=</span><span class="st">&quot; &quot;</span>)</span>
<span id="cb20-13"><a href="methods.html#cb20-13" aria-hidden="true" tabindex="-1"></a>        A <span class="op">=</span> img.copy()</span>
<span id="cb20-14"><a href="methods.html#cb20-14" aria-hidden="true" tabindex="-1"></a>        A[A<span class="op">!=</span>i]<span class="op">=</span><span class="dv">0</span></span>
<span id="cb20-15"><a href="methods.html#cb20-15" aria-hidden="true" tabindex="-1"></a>        _, B <span class="op">=</span> cv2.threshold(A, i<span class="op">-</span><span class="fl">0.5</span>, <span class="dv">255</span>, cv2.THRESH_BINARY)</span>
<span id="cb20-16"><a href="methods.html#cb20-16" aria-hidden="true" tabindex="-1"></a>        A,_ <span class="op">=</span> cv2.findContours(B.astype(np.uint8),cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)</span>
<span id="cb20-17"><a href="methods.html#cb20-17" aria-hidden="true" tabindex="-1"></a>        A <span class="op">=</span> pd.DataFrame(A[<span class="dv">0</span>].reshape(<span class="bu">len</span>(A[<span class="dv">0</span>]),<span class="dv">2</span>),columns<span class="op">=</span>[<span class="st">&#39;x_FOV_px&#39;</span>,<span class="st">&#39;y_FOV_px&#39;</span>])</span>
<span id="cb20-18"><a href="methods.html#cb20-18" aria-hidden="true" tabindex="-1"></a>        A[<span class="st">&#39;cell_ID&#39;</span>] <span class="op">=</span> <span class="st">&#39;c_1_</span><span class="sc">%d</span><span class="st">_</span><span class="sc">%d</span><span class="st">&#39;</span><span class="op">%</span>(fov,i)</span>
<span id="cb20-19"><a href="methods.html#cb20-19" aria-hidden="true" tabindex="-1"></a>        contours.append(A.copy())</span>
<span id="cb20-20"><a href="methods.html#cb20-20" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>()</span>
<span id="cb20-21"><a href="methods.html#cb20-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> pd.concat(contours)</span>
<span id="cb20-22"><a href="methods.html#cb20-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-23"><a href="methods.html#cb20-23" aria-hidden="true" tabindex="-1"></a>extractCellBoundaryPolygon(<span class="dv">2</span>).to_csv(<span class="st">&#39;Slide1_FOV002_cellPolygons.csv&#39;</span>,index<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb20-24"><a href="methods.html#cb20-24" aria-hidden="true" tabindex="-1"></a>extractCellBoundaryPolygon(<span class="dv">3</span>).to_csv(<span class="st">&#39;Slide1_FOV003_cellPolygons.csv&#39;</span>,index<span class="op">=</span><span class="va">False</span>)</span></code></pre></div>
<div id="the-function-to-read-one-fov-data" class="section level5 hasAnchor" number="3.6.2.3.1">
<h5><span class="header-section-number">3.6.2.3.1</span> The function to read one FOV data<a href="methods.html#the-function-to-read-one-fov-data" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<div class="sourceCode" id="cb21"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb21-1"><a href="methods.html#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os, re</span>
<span id="cb21-2"><a href="methods.html#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb21-3"><a href="methods.html#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb21-4"><a href="methods.html#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb21-5"><a href="methods.html#cb21-5" aria-hidden="true" tabindex="-1"></a>cellID_col <span class="op">=</span> <span class="st">&#39;cell_ID&#39;</span></span>
<span id="cb21-6"><a href="methods.html#cb21-6" aria-hidden="true" tabindex="-1"></a>suf_px<span class="op">=</span>{<span class="st">&#39;local&#39;</span>:<span class="st">&#39;_local_px&#39;</span>,<span class="st">&#39;global&#39;</span>:<span class="st">&#39;_global_px&#39;</span>}</span>
<span id="cb21-7"><a href="methods.html#cb21-7" aria-hidden="true" tabindex="-1"></a>sample_col <span class="op">=</span> <span class="st">&#39;library_id&#39;</span></span>
<span id="cb21-8"><a href="methods.html#cb21-8" aria-hidden="true" tabindex="-1"></a>cosMx_uns <span class="op">=</span> <span class="st">&#39;cosMx&#39;</span></span>
<span id="cb21-9"><a href="methods.html#cb21-9" aria-hidden="true" tabindex="-1"></a>cell_uns <span class="op">=</span> <span class="st">&#39;cell_polygons&#39;</span></span>
<span id="cb21-10"><a href="methods.html#cb21-10" aria-hidden="true" tabindex="-1"></a>transc_uns <span class="op">=</span> <span class="st">&#39;tx_loc&#39;</span></span>
<span id="cb21-11"><a href="methods.html#cb21-11" aria-hidden="true" tabindex="-1"></a>img_uns <span class="op">=</span> <span class="st">&#39;images&#39;</span></span>
<span id="cb21-12"><a href="methods.html#cb21-12" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> readOneCapture(strExp, <span class="co"># the path to the expression matrix of a FOV</span></span>
<span id="cb21-13"><a href="methods.html#cb21-13" aria-hidden="true" tabindex="-1"></a>                   strMeta, <span class="co"># the path to the cell meta table matrix of a FOV</span></span>
<span id="cb21-14"><a href="methods.html#cb21-14" aria-hidden="true" tabindex="-1"></a>                   strReduc, <span class="co"># a list of the paths to the expression matrix of a FOV</span></span>
<span id="cb21-15"><a href="methods.html#cb21-15" aria-hidden="true" tabindex="-1"></a>                   strTargetCoord, <span class="co"># the path to the target coordinate of a FOV</span></span>
<span id="cb21-16"><a href="methods.html#cb21-16" aria-hidden="true" tabindex="-1"></a>                   strCellBoundaries, <span class="co"># the path to the cell boundary polygons of a FOV</span></span>
<span id="cb21-17"><a href="methods.html#cb21-17" aria-hidden="true" tabindex="-1"></a>                   strImg, <span class="co"># the image associated with the FOV, like histology please make sure the scale/rotation matching the coordinates</span></span>
<span id="cb21-18"><a href="methods.html#cb21-18" aria-hidden="true" tabindex="-1"></a>                   sID, <span class="co">#the unique name of the FOV</span></span>
<span id="cb21-19"><a href="methods.html#cb21-19" aria-hidden="true" tabindex="-1"></a>                   cell_column<span class="op">=</span><span class="st">&#39;cell_id&#39;</span>,</span>
<span id="cb21-20"><a href="methods.html#cb21-20" aria-hidden="true" tabindex="-1"></a>                   lpx<span class="op">=</span><span class="st">&#39;_FOV_px&#39;</span>, <span class="co"># the column suffix which indicate the location on each FOV image in cell meta table</span></span>
<span id="cb21-21"><a href="methods.html#cb21-21" aria-hidden="true" tabindex="-1"></a>                   gpx<span class="op">=</span><span class="dv">4256</span> <span class="co"># the column suffix which indicate the location on the global of all FOV images in cell meta table. If a numeric, a fake global will be created vertically stack all FOVs</span></span>
<span id="cb21-22"><a href="methods.html#cb21-22" aria-hidden="true" tabindex="-1"></a>                  ):</span>
<span id="cb21-23"><a href="methods.html#cb21-23" aria-hidden="true" tabindex="-1"></a>    <span class="co"># exp</span></span>
<span id="cb21-24"><a href="methods.html#cb21-24" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot; exp&quot;</span>)</span>
<span id="cb21-25"><a href="methods.html#cb21-25" aria-hidden="true" tabindex="-1"></a>    gExp <span class="op">=</span> pd.read_csv(strExp,index_col<span class="op">=</span><span class="dv">0</span>).T</span>
<span id="cb21-26"><a href="methods.html#cb21-26" aria-hidden="true" tabindex="-1"></a>    gExp.columns.name<span class="op">=</span><span class="va">None</span></span>
<span id="cb21-27"><a href="methods.html#cb21-27" aria-hidden="true" tabindex="-1"></a>    <span class="co"># meta</span></span>
<span id="cb21-28"><a href="methods.html#cb21-28" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot; meta&quot;</span>)</span>
<span id="cb21-29"><a href="methods.html#cb21-29" aria-hidden="true" tabindex="-1"></a>    cInfo <span class="op">=</span> pd.read_csv(strMeta,index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb21-30"><a href="methods.html#cb21-30" aria-hidden="true" tabindex="-1"></a>    cID <span class="op">=</span> cInfo.index.intersection(gExp.index)</span>
<span id="cb21-31"><a href="methods.html#cb21-31" aria-hidden="true" tabindex="-1"></a>    <span class="co"># create anndata</span></span>
<span id="cb21-32"><a href="methods.html#cb21-32" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot; create AnnData&quot;</span>)</span>
<span id="cb21-33"><a href="methods.html#cb21-33" aria-hidden="true" tabindex="-1"></a>    D <span class="op">=</span> ad.AnnData(X<span class="op">=</span>gExp.loc[cID,:],obs<span class="op">=</span>cInfo.loc[cID,:])</span>
<span id="cb21-34"><a href="methods.html#cb21-34" aria-hidden="true" tabindex="-1"></a>    <span class="kw">del</span> gExp</span>
<span id="cb21-35"><a href="methods.html#cb21-35" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add histology image</span></span>
<span id="cb21-36"><a href="methods.html#cb21-36" aria-hidden="true" tabindex="-1"></a>    D.uns[cosMx_uns]<span class="op">=</span>{sID:{}}</span>
<span id="cb21-37"><a href="methods.html#cb21-37" aria-hidden="true" tabindex="-1"></a>    D.uns[cosMx_uns][sID][img_uns]<span class="op">=</span>plt.imread(strImg)</span>
<span id="cb21-38"><a href="methods.html#cb21-38" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add px coordinates</span></span>
<span id="cb21-39"><a href="methods.html#cb21-39" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot; add reduction&quot;</span>)</span>
<span id="cb21-40"><a href="methods.html#cb21-40" aria-hidden="true" tabindex="-1"></a>    D.obsm[<span class="st">&#39;X_FOV_local&#39;</span>] <span class="op">=</span> cInfo.loc[cID,[_ <span class="cf">for</span> _ <span class="kw">in</span> cInfo.columns <span class="cf">if</span> _.endswith(lpx)]].to_numpy(<span class="st">&#39;float32&#39;</span>)</span>
<span id="cb21-41"><a href="methods.html#cb21-41" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">isinstance</span>(gpx, (<span class="bu">int</span>, <span class="bu">float</span>)):</span>
<span id="cb21-42"><a href="methods.html#cb21-42" aria-hidden="true" tabindex="-1"></a>        FOV_global <span class="op">=</span> cInfo.loc[cID,[_ <span class="cf">for</span> _ <span class="kw">in</span> cInfo.columns <span class="cf">if</span> _.endswith(lpx)]]<span class="co">#</span></span>
<span id="cb21-43"><a href="methods.html#cb21-43" aria-hidden="true" tabindex="-1"></a>        FOV_global.iloc[:,<span class="dv">0</span>] <span class="op">+=</span> gpx</span>
<span id="cb21-44"><a href="methods.html#cb21-44" aria-hidden="true" tabindex="-1"></a>        FOV_global.iloc[:,<span class="dv">1</span>] <span class="op">=</span> D.uns[cosMx_uns][sID][img_uns].shape[<span class="dv">1</span>]<span class="op">-</span>FOV_global.iloc[:,<span class="dv">1</span>]<span class="op">-</span><span class="dv">1</span> <span class="co"># seems like the y(height) is flipped from the image and coordinates of transcript/cell boundaries</span></span>
<span id="cb21-45"><a href="methods.html#cb21-45" aria-hidden="true" tabindex="-1"></a>        D.obsm[<span class="st">&#39;X_FOV_global&#39;</span>] <span class="op">=</span> FOV_global.to_numpy(<span class="st">&#39;float32&#39;</span>)   </span>
<span id="cb21-46"><a href="methods.html#cb21-46" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb21-47"><a href="methods.html#cb21-47" aria-hidden="true" tabindex="-1"></a>        D.obsm[<span class="st">&#39;X_FOV_global&#39;</span>] <span class="op">=</span> cInfo.loc[cID,[_ <span class="cf">for</span> _ <span class="kw">in</span> cInfo.columns <span class="cf">if</span> _.endswith(lpx)]].to_numpy(<span class="st">&#39;float32&#39;</span>)    </span>
<span id="cb21-48"><a href="methods.html#cb21-48" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> oneF <span class="kw">in</span> strReduc:</span>
<span id="cb21-49"><a href="methods.html#cb21-49" aria-hidden="true" tabindex="-1"></a>        reducName <span class="op">=</span> re.sub(<span class="st">&quot;.csv&quot;</span>,<span class="st">&quot;&quot;</span>,os.path.basename(oneF).split(<span class="st">&quot;_&quot;</span>)[<span class="op">-</span><span class="dv">1</span>])</span>
<span id="cb21-50"><a href="methods.html#cb21-50" aria-hidden="true" tabindex="-1"></a>        D.obsm[<span class="st">&#39;X_&#39;</span><span class="op">+</span>reducName] <span class="op">=</span> pd.read_csv(oneF,index_col<span class="op">=</span><span class="dv">0</span>).loc[cID,].to_numpy(<span class="st">&#39;float32&#39;</span>)</span>
<span id="cb21-51"><a href="methods.html#cb21-51" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add cell boundaries</span></span>
<span id="cb21-52"><a href="methods.html#cb21-52" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot; add cell boundaries&quot;</span>)</span>
<span id="cb21-53"><a href="methods.html#cb21-53" aria-hidden="true" tabindex="-1"></a>    X <span class="op">=</span> pd.read_csv(strCellBoundaries)</span>
<span id="cb21-54"><a href="methods.html#cb21-54" aria-hidden="true" tabindex="-1"></a>    X.columns <span class="op">=</span> [re.sub(cell_column,cellID_col,re.sub(lpx,suf_px[<span class="st">&#39;local&#39;</span>],_)) <span class="cf">for</span> _ <span class="kw">in</span> X.columns]</span>
<span id="cb21-55"><a href="methods.html#cb21-55" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">isinstance</span>(gpx, (<span class="bu">int</span>, <span class="bu">float</span>)):</span>
<span id="cb21-56"><a href="methods.html#cb21-56" aria-hidden="true" tabindex="-1"></a>        X[<span class="st">&#39;x&#39;</span><span class="op">+</span>suf_px[<span class="st">&#39;global&#39;</span>]] <span class="op">=</span> X[<span class="st">&#39;x&#39;</span><span class="op">+</span>suf_px[<span class="st">&#39;local&#39;</span>]] <span class="op">+</span> gpx</span>
<span id="cb21-57"><a href="methods.html#cb21-57" aria-hidden="true" tabindex="-1"></a>        X[<span class="st">&#39;y&#39;</span><span class="op">+</span>suf_px[<span class="st">&#39;global&#39;</span>]] <span class="op">=</span> X[<span class="st">&#39;y&#39;</span><span class="op">+</span>suf_px[<span class="st">&#39;local&#39;</span>]]</span>
<span id="cb21-58"><a href="methods.html#cb21-58" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb21-59"><a href="methods.html#cb21-59" aria-hidden="true" tabindex="-1"></a>        X.columns <span class="op">=</span> [re.sub(gpx,suf_px[<span class="st">&#39;global&#39;</span>],_) <span class="cf">for</span> _ <span class="kw">in</span> X.columns]</span>
<span id="cb21-60"><a href="methods.html#cb21-60" aria-hidden="true" tabindex="-1"></a>    D.uns[cosMx_uns][sID][cell_uns] <span class="op">=</span> X    </span>
<span id="cb21-61"><a href="methods.html#cb21-61" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add transcript coordinates</span></span>
<span id="cb21-62"><a href="methods.html#cb21-62" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot; add transcript coordinates&quot;</span>)</span>
<span id="cb21-63"><a href="methods.html#cb21-63" aria-hidden="true" tabindex="-1"></a>    X <span class="op">=</span> pd.read_csv(strTargetCoord)</span>
<span id="cb21-64"><a href="methods.html#cb21-64" aria-hidden="true" tabindex="-1"></a>    X.columns <span class="op">=</span> [re.sub(cell_column,cellID_col,re.sub(lpx,suf_px[<span class="st">&#39;local&#39;</span>],_)) <span class="cf">for</span> _ <span class="kw">in</span> X.columns]</span>
<span id="cb21-65"><a href="methods.html#cb21-65" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">isinstance</span>(gpx, (<span class="bu">int</span>, <span class="bu">float</span>)):</span>
<span id="cb21-66"><a href="methods.html#cb21-66" aria-hidden="true" tabindex="-1"></a>        X[<span class="st">&#39;x&#39;</span><span class="op">+</span>suf_px[<span class="st">&#39;global&#39;</span>]] <span class="op">=</span> X[<span class="st">&#39;x&#39;</span><span class="op">+</span>suf_px[<span class="st">&#39;local&#39;</span>]] <span class="op">+</span> gpx</span>
<span id="cb21-67"><a href="methods.html#cb21-67" aria-hidden="true" tabindex="-1"></a>        X[<span class="st">&#39;y&#39;</span><span class="op">+</span>suf_px[<span class="st">&#39;global&#39;</span>]] <span class="op">=</span> X[<span class="st">&#39;y&#39;</span><span class="op">+</span>suf_px[<span class="st">&#39;local&#39;</span>]]</span>
<span id="cb21-68"><a href="methods.html#cb21-68" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb21-69"><a href="methods.html#cb21-69" aria-hidden="true" tabindex="-1"></a>        X.columns <span class="op">=</span> [re.sub(gpx,suf_px[<span class="st">&#39;global&#39;</span>],_) <span class="cf">for</span> _ <span class="kw">in</span> X.columns]</span>
<span id="cb21-70"><a href="methods.html#cb21-70" aria-hidden="true" tabindex="-1"></a>    D.uns[cosMx_uns][sID][transc_uns] <span class="op">=</span> X</span>
<span id="cb21-71"><a href="methods.html#cb21-71" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> D</span></code></pre></div>
</div>
<div id="use-two-fovs-from-human-liver-data-as-an-example" class="section level5 hasAnchor" number="3.6.2.3.2">
<h5><span class="header-section-number">3.6.2.3.2</span> Use two FOVs from human liver data as an example<a href="methods.html#use-two-fovs-from-human-liver-data-as-an-example" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<div class="sourceCode" id="cb22"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb22-1"><a href="methods.html#cb22-1" aria-hidden="true" tabindex="-1"></a>samples <span class="op">=</span> {}</span>
<span id="cb22-2"><a href="methods.html#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> [<span class="dv">2</span>,<span class="dv">3</span>]:</span>
<span id="cb22-3"><a href="methods.html#cb22-3" aria-hidden="true" tabindex="-1"></a>    samples[<span class="st">&#39;Normal</span><span class="sc">%d</span><span class="st">&#39;</span><span class="op">%</span>i] <span class="op">=</span> {}</span>
<span id="cb22-4"><a href="methods.html#cb22-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> one <span class="kw">in</span> [<span class="st">&#39;exp&#39;</span>,<span class="st">&#39;meta&#39;</span>,<span class="st">&#39;cellPolygons&#39;</span>,<span class="st">&#39;tx&#39;</span>,<span class="st">&#39;reduct&#39;</span>,<span class="st">&#39;img&#39;</span>]:</span>
<span id="cb22-5"><a href="methods.html#cb22-5" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> one <span class="op">==</span> <span class="st">&#39;reduct&#39;</span>:</span>
<span id="cb22-6"><a href="methods.html#cb22-6" aria-hidden="true" tabindex="-1"></a>            samples[<span class="st">&#39;Normal</span><span class="sc">%d</span><span class="st">&#39;</span><span class="op">%</span>i][one] <span class="op">=</span> [<span class="st">&#39;Slide1_FOV00</span><span class="sc">%d</span><span class="st">_UMAP.csv&#39;</span><span class="op">%</span>i,<span class="st">&#39;Slide1_FOV00</span><span class="sc">%d</span><span class="st">_slideUMAP.csv&#39;</span><span class="op">%</span>i]</span>
<span id="cb22-7"><a href="methods.html#cb22-7" aria-hidden="true" tabindex="-1"></a>        <span class="cf">elif</span> one <span class="op">==</span> <span class="st">&#39;img&#39;</span>:</span>
<span id="cb22-8"><a href="methods.html#cb22-8" aria-hidden="true" tabindex="-1"></a>            samples[<span class="st">&#39;Normal</span><span class="sc">%d</span><span class="st">&#39;</span><span class="op">%</span>i][one] <span class="op">=</span> <span class="st">&#39;NormalLiverFiles/CellStatsDir/CellComposite/CellComposite_F00</span><span class="sc">%d</span><span class="st">.jpg&#39;</span><span class="op">%</span>(i) <span class="co"># this can be replaced with the right histology image</span></span>
<span id="cb22-9"><a href="methods.html#cb22-9" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:<span class="co">#</span></span>
<span id="cb22-10"><a href="methods.html#cb22-10" aria-hidden="true" tabindex="-1"></a>            samples[<span class="st">&#39;Normal</span><span class="sc">%d</span><span class="st">&#39;</span><span class="op">%</span>i][one] <span class="op">=</span> <span class="st">&#39;Slide1_FOV00</span><span class="sc">%d</span><span class="st">_</span><span class="sc">%s</span><span class="st">.csv&#39;</span><span class="op">%</span>(i,one)</span>
<span id="cb22-11"><a href="methods.html#cb22-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-12"><a href="methods.html#cb22-12" aria-hidden="true" tabindex="-1"></a>adatas <span class="op">=</span> []</span>
<span id="cb22-13"><a href="methods.html#cb22-13" aria-hidden="true" tabindex="-1"></a>w <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb22-14"><a href="methods.html#cb22-14" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sID <span class="kw">in</span> samples.keys():</span>
<span id="cb22-15"><a href="methods.html#cb22-15" aria-hidden="true" tabindex="-1"></a>    D <span class="op">=</span> readOneCapture(samples[sID][<span class="st">&#39;exp&#39;</span>],</span>
<span id="cb22-16"><a href="methods.html#cb22-16" aria-hidden="true" tabindex="-1"></a>                       samples[sID][<span class="st">&#39;meta&#39;</span>],</span>
<span id="cb22-17"><a href="methods.html#cb22-17" aria-hidden="true" tabindex="-1"></a>                       samples[sID][<span class="st">&#39;reduct&#39;</span>],</span>
<span id="cb22-18"><a href="methods.html#cb22-18" aria-hidden="true" tabindex="-1"></a>                       samples[sID][<span class="st">&#39;tx&#39;</span>],</span>
<span id="cb22-19"><a href="methods.html#cb22-19" aria-hidden="true" tabindex="-1"></a>                       samples[sID][<span class="st">&#39;cellPolygons&#39;</span>],</span>
<span id="cb22-20"><a href="methods.html#cb22-20" aria-hidden="true" tabindex="-1"></a>                       samples[sID][<span class="st">&#39;img&#39;</span>],</span>
<span id="cb22-21"><a href="methods.html#cb22-21" aria-hidden="true" tabindex="-1"></a>                       sID,</span>
<span id="cb22-22"><a href="methods.html#cb22-22" aria-hidden="true" tabindex="-1"></a>                       gpx<span class="op">=</span>w</span>
<span id="cb22-23"><a href="methods.html#cb22-23" aria-hidden="true" tabindex="-1"></a>                      )</span>
<span id="cb22-24"><a href="methods.html#cb22-24" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If no global coordinates, a fake one align all of the input vertically</span></span>
<span id="cb22-25"><a href="methods.html#cb22-25" aria-hidden="true" tabindex="-1"></a>    w <span class="op">+=</span> D.uns[cosMx_uns][sID][img_uns].shape[<span class="dv">1</span>]</span>
<span id="cb22-26"><a href="methods.html#cb22-26" aria-hidden="true" tabindex="-1"></a>    adatas.append(D)</span>
<span id="cb22-27"><a href="methods.html#cb22-27" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> ad.AnnData.concatenate(<span class="op">*</span>adatas,</span>
<span id="cb22-28"><a href="methods.html#cb22-28" aria-hidden="true" tabindex="-1"></a>            join<span class="op">=</span><span class="st">&quot;outer&quot;</span>,</span>
<span id="cb22-29"><a href="methods.html#cb22-29" aria-hidden="true" tabindex="-1"></a>            batch_categories<span class="op">=</span><span class="bu">list</span>(samples.keys()),</span>
<span id="cb22-30"><a href="methods.html#cb22-30" aria-hidden="true" tabindex="-1"></a>            batch_key<span class="op">=</span>sample_col,</span>
<span id="cb22-31"><a href="methods.html#cb22-31" aria-hidden="true" tabindex="-1"></a>            index_unique<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb22-32"><a href="methods.html#cb22-32" aria-hidden="true" tabindex="-1"></a>            uns_merge<span class="op">=</span><span class="st">&#39;unique&#39;</span>)</span>
<span id="cb22-33"><a href="methods.html#cb22-33" aria-hidden="true" tabindex="-1"></a>adata.uns[cosMx_uns][<span class="st">&#39;keys&#39;</span>] <span class="op">=</span> {<span class="st">&#39;</span><span class="sc">%s</span><span class="st">_px&#39;</span><span class="op">%</span>i:{j:<span class="st">&#39;</span><span class="sc">%s%s</span><span class="st">&#39;</span><span class="op">%</span>(j,suf_px[i]) <span class="cf">for</span> j <span class="kw">in</span> [<span class="st">&#39;x&#39;</span>,<span class="st">&#39;y&#39;</span>]} <span class="cf">for</span> i <span class="kw">in</span> suf_px}</span>
<span id="cb22-34"><a href="methods.html#cb22-34" aria-hidden="true" tabindex="-1"></a>adata.uns[cosMx_uns][<span class="st">&#39;keys&#39;</span>][<span class="st">&#39;cell&#39;</span>] <span class="op">=</span> cell_uns</span>
<span id="cb22-35"><a href="methods.html#cb22-35" aria-hidden="true" tabindex="-1"></a>adata.uns[cosMx_uns][<span class="st">&#39;keys&#39;</span>][<span class="st">&#39;tx_loc&#39;</span>] <span class="op">=</span> transc_uns</span>
<span id="cb22-36"><a href="methods.html#cb22-36" aria-hidden="true" tabindex="-1"></a>adata.uns[cosMx_uns][<span class="st">&#39;keys&#39;</span>][<span class="st">&#39;img&#39;</span>] <span class="op">=</span> img_uns</span>
<span id="cb22-37"><a href="methods.html#cb22-37" aria-hidden="true" tabindex="-1"></a>adata.uns[cosMx_uns][<span class="st">&#39;keys&#39;</span>][<span class="st">&#39;tx_col&#39;</span>] <span class="op">=</span> <span class="st">&#39;target&#39;</span></span>
<span id="cb22-38"><a href="methods.html#cb22-38" aria-hidden="true" tabindex="-1"></a>adata.write(<span class="st">&quot;human_liver.h5ad&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="check-the-coordinates-among-cell-boundaries-transcript-locations-and-the-image-are-aligned" class="section level4 hasAnchor" number="3.6.2.4">
<h4><span class="header-section-number">3.6.2.4</span> Check the coordinates among cell boundaries, transcript locations and the image are aligned<a href="methods.html#check-the-coordinates-among-cell-boundaries-transcript-locations-and-the-image-are-aligned" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb23"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb23-1"><a href="methods.html#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> random</span>
<span id="cb23-2"><a href="methods.html#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb23-3"><a href="methods.html#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb23-4"><a href="methods.html#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb23-5"><a href="methods.html#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="co">#adata = ad.read_h5ad(&#39;human_liver.h5ad&#39;)</span></span>
<span id="cb23-6"><a href="methods.html#cb23-6" aria-hidden="true" tabindex="-1"></a>sID<span class="op">=</span> <span class="st">&#39;Normal3&#39;</span></span>
<span id="cb23-7"><a href="methods.html#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> showImg(img):</span>
<span id="cb23-8"><a href="methods.html#cb23-8" aria-hidden="true" tabindex="-1"></a>    plt.figure(figsize<span class="op">=</span>(<span class="dv">12</span>, <span class="dv">12</span>))</span>
<span id="cb23-9"><a href="methods.html#cb23-9" aria-hidden="true" tabindex="-1"></a>    plt.imshow(img,cmap<span class="op">=</span><span class="st">&#39;gray&#39;</span>)</span>
<span id="cb23-10"><a href="methods.html#cb23-10" aria-hidden="true" tabindex="-1"></a>    plt.axis(<span class="st">&#39;off&#39;</span>)</span>
<span id="cb23-11"><a href="methods.html#cb23-11" aria-hidden="true" tabindex="-1"></a>    plt.show()</span>
<span id="cb23-12"><a href="methods.html#cb23-12" aria-hidden="true" tabindex="-1"></a>    plt.close()</span>
<span id="cb23-13"><a href="methods.html#cb23-13" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> bresenham_line(pt1,pt2):</span>
<span id="cb23-14"><a href="methods.html#cb23-14" aria-hidden="true" tabindex="-1"></a>    x1,y1 <span class="op">=</span> pt1</span>
<span id="cb23-15"><a href="methods.html#cb23-15" aria-hidden="true" tabindex="-1"></a>    x2,y2 <span class="op">=</span> pt2</span>
<span id="cb23-16"><a href="methods.html#cb23-16" aria-hidden="true" tabindex="-1"></a>    points <span class="op">=</span> []</span>
<span id="cb23-17"><a href="methods.html#cb23-17" aria-hidden="true" tabindex="-1"></a>    dx <span class="op">=</span> <span class="bu">abs</span>(x2 <span class="op">-</span> x1)</span>
<span id="cb23-18"><a href="methods.html#cb23-18" aria-hidden="true" tabindex="-1"></a>    dy <span class="op">=</span> <span class="bu">abs</span>(y2 <span class="op">-</span> y1)</span>
<span id="cb23-19"><a href="methods.html#cb23-19" aria-hidden="true" tabindex="-1"></a>    sx <span class="op">=</span> <span class="dv">1</span> <span class="cf">if</span> x1 <span class="op">&lt;</span> x2 <span class="cf">else</span> <span class="op">-</span><span class="dv">1</span></span>
<span id="cb23-20"><a href="methods.html#cb23-20" aria-hidden="true" tabindex="-1"></a>    sy <span class="op">=</span> <span class="dv">1</span> <span class="cf">if</span> y1 <span class="op">&lt;</span> y2 <span class="cf">else</span> <span class="op">-</span><span class="dv">1</span></span>
<span id="cb23-21"><a href="methods.html#cb23-21" aria-hidden="true" tabindex="-1"></a>    error <span class="op">=</span> dx <span class="op">-</span> dy</span>
<span id="cb23-22"><a href="methods.html#cb23-22" aria-hidden="true" tabindex="-1"></a>    <span class="cf">while</span> x1 <span class="op">!=</span> x2 <span class="kw">or</span> y1 <span class="op">!=</span> y2:</span>
<span id="cb23-23"><a href="methods.html#cb23-23" aria-hidden="true" tabindex="-1"></a>        points.append((x1, y1))</span>
<span id="cb23-24"><a href="methods.html#cb23-24" aria-hidden="true" tabindex="-1"></a>        e2 <span class="op">=</span> <span class="dv">2</span> <span class="op">*</span> error</span>
<span id="cb23-25"><a href="methods.html#cb23-25" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> e2 <span class="op">&gt;</span> <span class="op">-</span>dy:</span>
<span id="cb23-26"><a href="methods.html#cb23-26" aria-hidden="true" tabindex="-1"></a>            error <span class="op">-=</span> dy</span>
<span id="cb23-27"><a href="methods.html#cb23-27" aria-hidden="true" tabindex="-1"></a>            x1 <span class="op">+=</span> sx</span>
<span id="cb23-28"><a href="methods.html#cb23-28" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> e2 <span class="op">&lt;</span> dx:</span>
<span id="cb23-29"><a href="methods.html#cb23-29" aria-hidden="true" tabindex="-1"></a>            error <span class="op">+=</span> dx</span>
<span id="cb23-30"><a href="methods.html#cb23-30" aria-hidden="true" tabindex="-1"></a>            y1 <span class="op">+=</span> sy</span>
<span id="cb23-31"><a href="methods.html#cb23-31" aria-hidden="true" tabindex="-1"></a>    points.append((x2, y2))</span>
<span id="cb23-32"><a href="methods.html#cb23-32" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> points</span>
<span id="cb23-33"><a href="methods.html#cb23-33" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> loc_img(pt,imgH):</span>
<span id="cb23-34"><a href="methods.html#cb23-34" aria-hidden="true" tabindex="-1"></a>    x,y <span class="op">=</span> pt</span>
<span id="cb23-35"><a href="methods.html#cb23-35" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span>((<span class="bu">round</span>(y),<span class="bu">round</span>(x)))<span class="co">#imgH-y-1</span></span>
<span id="cb23-36"><a href="methods.html#cb23-36" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> square_integer(ct,halfS):</span>
<span id="cb23-37"><a href="methods.html#cb23-37" aria-hidden="true" tabindex="-1"></a>    x_center, y_center <span class="op">=</span> ct</span>
<span id="cb23-38"><a href="methods.html#cb23-38" aria-hidden="true" tabindex="-1"></a>    coordinates_in_square <span class="op">=</span> []</span>
<span id="cb23-39"><a href="methods.html#cb23-39" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> x <span class="kw">in</span> <span class="bu">range</span>(x_center <span class="op">-</span> halfS, x_center <span class="op">+</span> halfS <span class="op">+</span> <span class="dv">1</span>):</span>
<span id="cb23-40"><a href="methods.html#cb23-40" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> y <span class="kw">in</span> <span class="bu">range</span>(y_center <span class="op">-</span> halfS, y_center <span class="op">+</span> halfS <span class="op">+</span> <span class="dv">1</span>):</span>
<span id="cb23-41"><a href="methods.html#cb23-41" aria-hidden="true" tabindex="-1"></a>            coordinates_in_square.append((x, y))</span>
<span id="cb23-42"><a href="methods.html#cb23-42" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> coordinates_in_square</span>
<span id="cb23-43"><a href="methods.html#cb23-43" aria-hidden="true" tabindex="-1"></a>cosMxKey<span class="op">=</span><span class="st">&#39;cosMx&#39;</span></span>
<span id="cb23-44"><a href="methods.html#cb23-44" aria-hidden="true" tabindex="-1"></a>keys <span class="op">=</span> adata.uns[cosMxKey][<span class="st">&#39;keys&#39;</span>]</span>
<span id="cb23-45"><a href="methods.html#cb23-45" aria-hidden="true" tabindex="-1"></a>cosMxArray<span class="op">=</span>{}</span>
<span id="cb23-46"><a href="methods.html#cb23-46" aria-hidden="true" tabindex="-1"></a>cood_pair<span class="op">=</span>{}</span>
<span id="cb23-47"><a href="methods.html#cb23-47" aria-hidden="true" tabindex="-1"></a>imgC <span class="op">=</span> adata.uns[cosMxKey][sID][keys[<span class="st">&#39;img&#39;</span>]].copy()</span>
<span id="cb23-48"><a href="methods.html#cb23-48" aria-hidden="true" tabindex="-1"></a>cell_color<span class="op">=</span> <span class="st">&#39;ffffff&#39;</span> <span class="co">#&#39;000000&#39;</span></span>
<span id="cb23-49"><a href="methods.html#cb23-49" aria-hidden="true" tabindex="-1"></a>rgb <span class="op">=</span> np.array(<span class="bu">tuple</span>(<span class="bu">int</span>(cell_color[i:i<span class="op">+</span><span class="dv">2</span>], <span class="dv">16</span>) <span class="cf">for</span> i <span class="kw">in</span> (<span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">4</span>)),dtype<span class="op">=</span><span class="st">&#39;uint8&#39;</span>)</span>
<span id="cb23-50"><a href="methods.html#cb23-50" aria-hidden="true" tabindex="-1"></a>cellB <span class="op">=</span> adata.uns[cosMxKey][sID][keys[<span class="st">&#39;cell&#39;</span>]]</span>
<span id="cb23-51"><a href="methods.html#cb23-51" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> cellB.cell_ID.unique():</span>
<span id="cb23-52"><a href="methods.html#cb23-52" aria-hidden="true" tabindex="-1"></a>    oneC <span class="op">=</span> cellB[cellB.cell_ID<span class="op">==</span>i].reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb23-53"><a href="methods.html#cb23-53" aria-hidden="true" tabindex="-1"></a>    N <span class="op">=</span> oneC.shape[<span class="dv">0</span>]</span>
<span id="cb23-54"><a href="methods.html#cb23-54" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> j <span class="kw">in</span> <span class="bu">range</span>(N):</span>
<span id="cb23-55"><a href="methods.html#cb23-55" aria-hidden="true" tabindex="-1"></a>        p <span class="op">=</span> bresenham_line(loc_img((oneC[keys[<span class="st">&#39;local_px&#39;</span>][<span class="st">&#39;x&#39;</span>]][j<span class="op">%</span>N],oneC[keys[<span class="st">&#39;local_px&#39;</span>][<span class="st">&#39;y&#39;</span>]][j<span class="op">%</span>N]),imgC.shape[<span class="dv">1</span>]),</span>
<span id="cb23-56"><a href="methods.html#cb23-56" aria-hidden="true" tabindex="-1"></a>                           loc_img((oneC[keys[<span class="st">&#39;local_px&#39;</span>][<span class="st">&#39;x&#39;</span>]][(j<span class="op">+</span><span class="dv">1</span>)<span class="op">%</span>N],oneC[keys[<span class="st">&#39;local_px&#39;</span>][<span class="st">&#39;y&#39;</span>]][(j<span class="op">+</span><span class="dv">1</span>)<span class="op">%</span>N]),imgC.shape[<span class="dv">1</span>]))</span>
<span id="cb23-57"><a href="methods.html#cb23-57" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> a <span class="kw">in</span> p:</span>
<span id="cb23-58"><a href="methods.html#cb23-58" aria-hidden="true" tabindex="-1"></a>            imgC[a]<span class="op">=</span>rgb</span>
<span id="cb23-59"><a href="methods.html#cb23-59" aria-hidden="true" tabindex="-1"></a>showImg(imgC)</span>
<span id="cb23-60"><a href="methods.html#cb23-60" aria-hidden="true" tabindex="-1"></a><span class="co"># cell boundries with transcript within the same randomly selected 10 cell</span></span>
<span id="cb23-61"><a href="methods.html#cb23-61" aria-hidden="true" tabindex="-1"></a>imgC <span class="op">=</span> np.ones((adata.uns[cosMxKey][sID][keys[<span class="st">&#39;img&#39;</span>]].shape[<span class="dv">0</span>],adata.uns[cosMxKey][sID][keys[<span class="st">&#39;img&#39;</span>]].shape[<span class="dv">1</span>],<span class="dv">3</span>),dtype<span class="op">=</span><span class="st">&#39;uint8&#39;</span>)<span class="op">*</span><span class="dv">255</span></span>
<span id="cb23-62"><a href="methods.html#cb23-62" aria-hidden="true" tabindex="-1"></a>cell_color<span class="op">=</span><span class="st">&quot;000000&quot;</span></span>
<span id="cb23-63"><a href="methods.html#cb23-63" aria-hidden="true" tabindex="-1"></a>rgb <span class="op">=</span> np.array(<span class="bu">tuple</span>(<span class="bu">int</span>(cell_color[i:i<span class="op">+</span><span class="dv">2</span>], <span class="dv">16</span>) <span class="cf">for</span> i <span class="kw">in</span> (<span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">4</span>)),dtype<span class="op">=</span><span class="st">&#39;uint8&#39;</span>)</span>
<span id="cb23-64"><a href="methods.html#cb23-64" aria-hidden="true" tabindex="-1"></a>cID <span class="op">=</span> random.sample(<span class="bu">list</span>(cellB.cell_ID.unique()),<span class="dv">10</span>)</span>
<span id="cb23-65"><a href="methods.html#cb23-65" aria-hidden="true" tabindex="-1"></a>txLoc <span class="op">=</span> adata.uns[<span class="st">&#39;cosMx&#39;</span>][sID][keys[<span class="st">&#39;tx_loc&#39;</span>]]</span>
<span id="cb23-66"><a href="methods.html#cb23-66" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> cID:</span>
<span id="cb23-67"><a href="methods.html#cb23-67" aria-hidden="true" tabindex="-1"></a>    oneC <span class="op">=</span> cellB[cellB.cell_ID<span class="op">==</span>i].reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb23-68"><a href="methods.html#cb23-68" aria-hidden="true" tabindex="-1"></a>    N <span class="op">=</span> oneC.shape[<span class="dv">0</span>]</span>
<span id="cb23-69"><a href="methods.html#cb23-69" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> j <span class="kw">in</span> <span class="bu">range</span>(N):</span>
<span id="cb23-70"><a href="methods.html#cb23-70" aria-hidden="true" tabindex="-1"></a>        p <span class="op">=</span> bresenham_line(loc_img((oneC[keys[<span class="st">&#39;local_px&#39;</span>][<span class="st">&#39;x&#39;</span>]][j<span class="op">%</span>N],oneC[keys[<span class="st">&#39;local_px&#39;</span>][<span class="st">&#39;y&#39;</span>]][j<span class="op">%</span>N]),imgC.shape[<span class="dv">1</span>]),</span>
<span id="cb23-71"><a href="methods.html#cb23-71" aria-hidden="true" tabindex="-1"></a>                           loc_img((oneC[keys[<span class="st">&#39;local_px&#39;</span>][<span class="st">&#39;x&#39;</span>]][(j<span class="op">+</span><span class="dv">1</span>)<span class="op">%</span>N],oneC[keys[<span class="st">&#39;local_px&#39;</span>][<span class="st">&#39;y&#39;</span>]][(j<span class="op">+</span><span class="dv">1</span>)<span class="op">%</span>N]),imgC.shape[<span class="dv">1</span>]))</span>
<span id="cb23-72"><a href="methods.html#cb23-72" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> a <span class="kw">in</span> p:</span>
<span id="cb23-73"><a href="methods.html#cb23-73" aria-hidden="true" tabindex="-1"></a>            imgC[a]<span class="op">=</span>rgb</span>
<span id="cb23-74"><a href="methods.html#cb23-74" aria-hidden="true" tabindex="-1"></a>col <span class="op">=</span> <span class="st">&#39;aa0000&#39;</span></span>
<span id="cb23-75"><a href="methods.html#cb23-75" aria-hidden="true" tabindex="-1"></a>rgb <span class="op">=</span> np.array(<span class="bu">tuple</span>(<span class="bu">int</span>(col[i:i<span class="op">+</span><span class="dv">2</span>], <span class="dv">16</span>) <span class="cf">for</span> i <span class="kw">in</span> (<span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">4</span>)),dtype<span class="op">=</span><span class="st">&#39;uint8&#39;</span>)</span>
<span id="cb23-76"><a href="methods.html#cb23-76" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> txLoc.index[txLoc[cellID_col].isin(cID)].tolist():</span>
<span id="cb23-77"><a href="methods.html#cb23-77" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> a <span class="kw">in</span> square_integer(loc_img((txLoc[keys[<span class="st">&#39;local_px&#39;</span>][<span class="st">&#39;x&#39;</span>]][i],txLoc[keys[<span class="st">&#39;local_px&#39;</span>][<span class="st">&#39;y&#39;</span>]][i]),imgC.shape[<span class="dv">1</span>]),<span class="dv">4</span>):</span>
<span id="cb23-78"><a href="methods.html#cb23-78" aria-hidden="true" tabindex="-1"></a>            imgC[a] <span class="op">=</span> rgb</span>
<span id="cb23-79"><a href="methods.html#cb23-79" aria-hidden="true" tabindex="-1"></a>showImg(imgC) </span></code></pre></div>

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="how-to-use-cellxgene-vip.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="helpful-tips.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["cellxgene_VIP.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
